<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zining Qi" />


<title>Initial processing and exploration of the autoimmune GWAS data</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ebmf_immune</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zq2209/ebmf_immune">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Initial processing and exploration of the
autoimmune GWAS data</h1>
<h4 class="author">Zining Qi</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>ebmf_immune/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree30a211dc8958e5c48fc86c3e53831a62ed27ac07targetblank30a211da">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/zq2209/ebmf_immune/tree/30a211dc8958e5c48fc86c3e53831a62ed27ac07" target="_blank">30a211d</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree30a211dc8958e5c48fc86c3e53831a62ed27ac07targetblank30a211da"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/zq2209/ebmf_immune/tree/30a211dc8958e5c48fc86c3e53831a62ed27ac07" target="_blank">30a211d</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

Unstaged changes:
    Modified:   analysis/autoimmune_factors.Rmd

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/autoimmune.Rmd</code>) and HTML
(<code>docs/autoimmune.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/30a211dc8958e5c48fc86c3e53831a62ed27ac07/analysis/autoimmune.Rmd" target="_blank">30a211d</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/zq2209/ebmf_immune/86a0ff284c88aad27c8d1f997733714f0e68d9af/docs/autoimmune.html" target="_blank">86a0ff2</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/analysis/autoimmune.Rmd" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/zq2209/ebmf_immune/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/autoimmune.html" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/4a0f61c7237aae19aa2c8350e0929b447018c369/analysis/autoimmune.Rmd" target="_blank">4a0f61c</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/ef8b718330f0543c05d457c3f7fbc2cf94e64892/analysis/autoimmune.Rmd" target="_blank">ef8b718</a>
</td>
<td>
Zining Qi
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/865d316b479332a8baac65333fa6e196f03dd2bd/analysis/autoimmune.Rmd" target="_blank">865d316</a>
</td>
<td>
Zining Qi
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This analysis processes genome-wide association study (GWAS) data for
multiple autoimmune diseases obtained from the <a
href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>. The goal is to
prepare the data for subsequent analyses by:</p>
<ol style="list-style-type: decimal">
<li>Identifying variants common across all traits</li>
<li>Extracting effect sizes (<span class="math inline">\(\beta\)</span>)
and standard errors</li>
<li>Cleaning and organizing the data</li>
<li>Visualizing distributions</li>
<li>Creating analysis-ready matrices</li>
</ol>
<p>The complete list of autoimmune diseases included can be found <a
href="https://docs.google.com/spreadsheets/d/1SmX4JMabPz-nPkBrG63zJNYbph50JFiJ15y6fThV3So/edit?gid=0#gid=0">here</a>.</p>
<div id="setup" class="section level3">
<h3>Setup</h3>
<p>First, we load required packages and set knitr options:</p>
<pre class="r"><code># Load required packages
library(tidyverse)    # For data manipulation and visualization
library(data.table)   # For efficient data reading and processing
library(pheatmap)</code></pre>
</div>
<div id="identify-common-variants-across-studies"
class="section level3">
<h3>Identify Common Variants Across Studies</h3>
<p>This step is to find genetic variants that are present in all GWAS
studies to enable cross-study comparisons.</p>
<pre class="r"><code># List all GWAS data files
files &lt;- list.files(pattern = &quot;\\.h\\.tsv$&quot;)

# Initialize with variants from first file
common_variants &lt;- fread(files[1], select = &quot;variant_id&quot;)[[1]]

# Iterate through files to find intersection of variants
for (file in files[c(-1, -12, -13)]) {
  current_variants &lt;- fread(file, select = &quot;variant_id&quot;)[[1]]
  common_variants &lt;- intersect(common_variants, current_variants)
  cat(&quot;Processed&quot;, file, &quot;| Remaining variants:&quot;, length(common_variants), &quot;\n&quot;)
}

# Save results
writeLines(common_variants, &quot;common_variant_ids.txt&quot;)
cat(&quot;Final number of overlapping variants:&quot;, length(common_variants))</code></pre>
<p>Final number of overlapping variants are over 18000. Two traits were
dropped because the overlapping of these two traits will be too low.</p>
<p><strong>Fixme</strong>:</p>
<ol style="list-style-type: decimal">
<li>The number of overlapping is low?</li>
<li>No stringent processing on variants now, such as no MAF filter,
harmonized of effects.</li>
</ol>
</div>
<div id="extract-effect-sizes-for-common-variants"
class="section level3">
<h3>Extract Effect Sizes for Common Variants</h3>
<p>For each study, extract effect sizes (<span
class="math inline">\(\beta\)</span>) only for the common variants
identified in Step 1.</p>
<pre class="r"><code>results_list &lt;- list()
duplicate_report &lt;- list()

for (file in files[c(-12, -13)]) {
  # Read relevant columns
  gwas_data &lt;- fread(file, select = c(&quot;variant_id&quot;, &quot;beta&quot;, &quot;standard_error&quot;))
  
  # Filter to common variants and compute Z-scores
  filtered_data &lt;- gwas_data %&gt;% 
    filter(variant_id %in% common_variants) %&gt;%
    mutate(
      source = gsub(&quot;\\.h\\.tsv$&quot;, &quot;&quot;, file),
      z_hat = beta / standard_error
    )
  
  # Identify and report duplicates
  dup_variants &lt;- filtered_data$variant_id[duplicated(filtered_data$variant_id)]
  if(length(dup_variants) &gt; 0) {
    duplicate_report[[file]] &lt;- data.table(
      file = file,
      variant_id = unique(dup_variants),
      n_duplicates = table(filtered_data$variant_id)[unique(dup_variants)]
    )
  }
  
  # Keep only first occurrence of each variant
  results_list[[file]] &lt;- filtered_data[!duplicated(filtered_data$variant_id), ]
}

# Combine all studies
combined_results &lt;- bind_rows(results_list)</code></pre>
</div>
<div id="map-study-ids-to-disease-names" class="section level3">
<h3>Map Study IDs to Disease Names</h3>
<p>Then, replace study identifiers with meaningful disease names for
clearer interpretation.</p>
<pre class="r"><code># Create study-to-disease mapping
study_to_disease &lt;- data.frame(
  source = c(&quot;20190752-GCST000612-EFO_0001060&quot;,
             &quot;23749187-GCST005529-EFO_0003898&quot;,
             &quot;26192919-GCST003044-EFO_0000384&quot;,
             &quot;27329760-GCST003724-EFO_0000289&quot;,
             &quot;23143594-GCST005527-EFO_0000676&quot;,
             &quot;24076602-GCST005531-EFO_0003885&quot;,
             &quot;26192919-GCST003045-EFO_0000729&quot;,
             &quot;33830302-GCST90000529-EFO_0001359&quot;,
             &quot;23143596-GCST005569-EFO_0000685&quot;,
             &quot;26192919-GCST003043-EFO_0003767&quot;,
             &quot;26502338-GCST003156-EFO_0002690&quot;),
  disease = c(&quot;celiac_disease&quot;,
              &quot;ankylosing_spondylitis&quot;,
              &quot;crohns_disease&quot;,
              &quot;biopolar_disorder&quot;,
              &quot;psoriasis&quot;,
              &quot;multiple_sclerosis&quot;,
              &quot;ulcerative_colitis&quot;,
              &quot;type_1_diabetes&quot;,
              &quot;rheumatoid_arthritis&quot;,
              &quot;inflammatory_bowel_disease&quot;,
              &quot;systemic_lupus_erythematosus&quot;)
)

# Merge mapping with results
combined_results &lt;- merge(combined_results, study_to_disease, by = &quot;source&quot;, all.x = TRUE)

# Save processed data
save(combined_results, file = &#39;../data/autoimmune_combined_results.RData&#39;)</code></pre>
</div>
<div id="visualize-effect-size-distributions" class="section level3">
<h3>Visualize Effect Size Distributions</h3>
<div id="overall-distribution-of-effect-sizes-beta"
class="section level4">
<h4>Overall Distribution of Effect Sizes (<span
class="math inline">\(\beta\)</span>)</h4>
<p>Before running any matrix facotorization, we want to examine the
distribution of genetic effect sizes across all studies.</p>
<pre class="r"><code>load(&#39;../data/autoimmune_combined_results.RData&#39;)</code></pre>
<pre class="r"><code>ggplot(combined_results, aes(x = beta)) +
  geom_histogram(binwidth = 0.01, fill = &quot;#21908CFF&quot;, color = &quot;black&quot;) +
  theme_minimal() +
  labs(x = expression(hat(beta)),
       y = &quot;Count&quot;)</code></pre>
<p><img src="figure/autoimmune.Rmd/beta-distribution-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-beta-distribution-1">
Past versions of beta-distribution-1.png
</button>
</p>
<div id="fig-beta-distribution-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/figure/autoimmune.Rmd/beta-distribution-1.png" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="effect-size-distribution-by-disease" class="section level4">
<h4>Effect Size Distribution by Disease</h4>
<p>Then, compare effect size distributions across different autoimmune
diseases.</p>
<pre class="r"><code>ggplot(combined_results, aes(x = beta)) +
  geom_histogram(binwidth = 0.03, fill = &quot;#21908CFF&quot;, color = &quot;black&quot;) +
  facet_wrap(~ disease, scales = &quot;free_y&quot;) +
  theme_minimal() +
  labs(x = expression(hat(beta)),
       y = &quot;Count&quot;)</code></pre>
<p><img src="figure/autoimmune.Rmd/beta-by-disease-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-beta-by-disease-1">
Past versions of beta-by-disease-1.png
</button>
</p>
<div id="fig-beta-by-disease-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/figure/autoimmune.Rmd/beta-by-disease-1.png" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="overall-distribution-of-z-scores" class="section level4">
<h4>Overall Distribution of Z-scores</h4>
<p>Additional, we examine the distribution of Z-scores (<span
class="math inline">\(\beta\)</span>/SE) which indicate significance of
associations.</p>
<pre class="r"><code>ggplot(combined_results, aes(x = z_hat)) +
  geom_histogram(binwidth = 1, fill = &quot;#21908CFF&quot;, color = &quot;black&quot;) +
  theme_minimal() +
  labs(x = expression(hat(Z)),
       y = &quot;Count&quot;)
# Warning: Removed 2369 rows containing non-finite outside the scale range
# (`stat_bin()`).</code></pre>
<p><img src="figure/autoimmune.Rmd/zscore-distribution-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-zscore-distribution-1">
Past versions of zscore-distribution-1.png
</button>
</p>
<div id="fig-zscore-distribution-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/figure/autoimmune.Rmd/zscore-distribution-1.png" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="z-score-distribution-by-disease" class="section level4">
<h4>Z-score Distribution by Disease</h4>
<p>Also, compare Z-score distributions across different autoimmune
diseases.</p>
<pre class="r"><code>ggplot(combined_results, aes(x = z_hat)) +
  geom_histogram(binwidth = 1, fill = &quot;#21908CFF&quot;, color = &quot;black&quot;) +
  facet_wrap(~ disease, scales = &quot;free_y&quot;) +
  theme_minimal() +
  labs(x = expression(hat(Z)),
       y = &quot;Count&quot;)
# Warning: Removed 2369 rows containing non-finite outside the scale range
# (`stat_bin()`).</code></pre>
<p><img src="figure/autoimmune.Rmd/zscore-by-disease-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-zscore-by-disease-1">
Past versions of zscore-by-disease-1.png
</button>
</p>
<div id="fig-zscore-by-disease-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/figure/autoimmune.Rmd/zscore-by-disease-1.png" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><strong>Fixme</strong>:</p>
<ol style="list-style-type: decimal">
<li>Beta distribution seems normal right now but effect size are small,
use z score instead?</li>
</ol>
</div>
</div>
<div id="create-analysis-ready-matrices" class="section level3">
<h3>Create Analysis-Ready Matrices</h3>
<p>This step is to reshape data into variant-by-disease matrices for
downstream analysis.</p>
<pre class="r"><code># Create variant-by-disease beta matrix
beta_matrix &lt;- combined_results %&gt;%
  pivot_wider(
    id_cols = variant_id,
    names_from = disease,
    values_from = beta
  ) %&gt;% 
  as.data.frame() %&gt;%
  column_to_rownames(&quot;variant_id&quot;) 

z_matrix &lt;- combined_results %&gt;%
  pivot_wider(
    id_cols = variant_id,
    names_from = disease,
    values_from = z_hat
  ) %&gt;% 
  as.data.frame() %&gt;%
  column_to_rownames(&quot;variant_id&quot;) 

# Save in multiple formats
#save(beta_matrix, file = &quot;beta_matrix.RData&quot;)
#fwrite(beta_matrix, row.names = TRUE, file = &quot;beta_matrix.csv&quot;)</code></pre>
</div>
<div id="visualize-patterns-in-effect-sizes" class="section level3">
<h3>Visualize Patterns in Effect Sizes</h3>
<p>Fianlly, we can explore patterns and relationships between diseases
through heatmap visualization.</p>
<pre class="r"><code># Heatmap of effect sizes (scaled by row)
pheatmap(beta_matrix, 
         cluster_cols = FALSE, 
         scale = &#39;row&#39;, 
         cluster_rows = FALSE,
         show_colnames = TRUE,
         show_rownames = FALSE,
         color = colorRampPalette(c(&quot;#440154FF&quot;, &quot;white&quot;, &quot;#21908CFF&quot;))(100)
)</code></pre>
<p><img src="figure/autoimmune.Rmd/heatmap-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-heatmap-1">
Past versions of heatmap-1.png
</button>
</p>
<div id="fig-heatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/d287a6767c0a68e4b55a6e3af802c18ed455ddaa/docs/figure/autoimmune.Rmd/heatmap-1.png" target="_blank">d287a67</a>
</td>
<td>
“John
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>
# pheatmap(z_matrix, 
#          cluster_cols = FALSE, 
#          scale = &#39;row&#39;, 
#          cluster_rows = FALSE,
#          show_colnames = TRUE,
#          show_rownames = FALSE,
#          color = colorRampPalette(c(&quot;#440154FF&quot;, &quot;white&quot;, &quot;#21908CFF&quot;))(100)
# )</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.4.3 (2025-02-28)
# Platform: aarch64-apple-darwin20
# Running under: macOS Sequoia 15.4.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] pheatmap_1.0.12   data.table_1.17.0 lubridate_1.9.4   forcats_1.0.0    
#  [5] stringr_1.5.1     dplyr_1.1.4       purrr_1.0.4       readr_2.1.5      
#  [9] tidyr_1.3.1       tibble_3.2.1      ggplot2_3.5.2     tidyverse_2.0.0  
# 
# loaded via a namespace (and not attached):
#  [1] sass_0.4.10        generics_0.1.3     stringi_1.8.7      hms_1.1.3         
#  [5] digest_0.6.37      magrittr_2.0.3     evaluate_1.0.3     grid_4.4.3        
#  [9] timechange_0.3.0   RColorBrewer_1.1-3 fastmap_1.2.0      rprojroot_2.0.4   
# [13] workflowr_1.7.1    jsonlite_2.0.0     whisker_0.4.1      promises_1.3.2    
# [17] scales_1.4.0       jquerylib_0.1.4    cli_3.6.5          rlang_1.1.6       
# [21] withr_3.0.2        cachem_1.1.0       yaml_2.3.10        tools_4.4.3       
# [25] tzdb_0.5.0         httpuv_1.6.15      vctrs_0.6.5        R6_2.6.1          
# [29] lifecycle_1.0.4    git2r_0.36.2       fs_1.6.6           pkgconfig_2.0.3   
# [33] pillar_1.10.2      bslib_0.9.0        later_1.4.2        gtable_0.3.6      
# [37] glue_1.8.0         Rcpp_1.0.14        xfun_0.52          tidyselect_1.2.1  
# [41] rstudioapi_0.17.1  knitr_1.50         farver_2.1.2       htmltools_0.5.8.1 
# [45] labeling_0.4.3     rmarkdown_2.29     compiler_4.4.3</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
