<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zining Qi" />


<title>Matrix factorization of significant signals</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ebmf_immune</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zq2209/ebmf_immune">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Matrix factorization of significant
signals</h1>
<h4 class="author">Zining Qi</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-06-09
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>ebmf_immune/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree3a315e9e5b846a484f2f232b91aaf04f3f0d5d67targetblank3a315e9a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/zq2209/ebmf_immune/tree/3a315e9e5b846a484f2f232b91aaf04f3f0d5d67" target="_blank">3a315e9</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree3a315e9e5b846a484f2f232b91aaf04f3f0d5d67targetblank3a315e9a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/zq2209/ebmf_immune/tree/3a315e9e5b846a484f2f232b91aaf04f3f0d5d67" target="_blank">3a315e9</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rproj.user/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown
(<code>analysis/autoimmune_factors_significant.Rmd</code>) and HTML
(<code>docs/autoimmune_factors_significant.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/zq2209/ebmf_immune/9fba759b8c919eaedca7f9ea1af24a1da0a41091/docs/autoimmune_factors_significant.html" target="_blank">9fba759</a>
</td>
<td>
zq2209
</td>
<td>
2025-06-09
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/0c8ed180ec60803002fb31565c8ac8badb6c1503/analysis/autoimmune_factors_significant.Rmd" target="_blank">0c8ed18</a>
</td>
<td>
zq2209
</td>
<td>
2025-06-09
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/260a7828490ee3f4c0d5f182002e13a351c803d5/analysis/autoimmune_factors_significant.Rmd" target="_blank">260a782</a>
</td>
<td>
“John
</td>
<td>
2025-06-08
</td>
<td>
update analysis on significant signals
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This analysis processes genome-wide association study (GWAS) data for
multiple autoimmune diseases obtained from the <a
href="https://www.ebi.ac.uk/gwas/">GWAS Catalog</a>. The goal is to
prepare the data for subsequent analyses by:</p>
<ol style="list-style-type: decimal">
<li>Identifying variants common across all traits</li>
<li>Extracting effect sizes (<span class="math inline">\(\beta\)</span>)
and standard errors</li>
<li>Cleaning and organizing the data</li>
<li>Visualizing distributions</li>
<li>Creating analysis-ready matrices</li>
</ol>
<p>The complete list of autoimmune diseases included can be found <a
href="https://docs.google.com/spreadsheets/d/1SmX4JMabPz-nPkBrG63zJNYbph50JFiJ15y6fThV3So/edit?gid=0#gid=0">here</a>.</p>
<div id="setup" class="section level3">
<h3>Setup</h3>
<p>First, we load required packages and set knitr options:</p>
<pre class="r"><code># Load required packages
library(tidyverse)    # For data manipulation and visualization
library(data.table)   # For efficient data reading and processing
library(pheatmap)
library(forestplot)
library(png)
library(grid)
library(gridExtra)

# Data manipulation and visualization
library(tidyverse)
library(reshape2)
library(cowplot)
library(viridis)

# Matrix factorization tools
library(flashier)
library(fastTopics)
</code></pre>
</div>
<div id="identify-significant-variants-across-studies"
class="section level3">
<h3>Identify Significant Variants Across Studies</h3>
<p>This step is to find genetic variants that are present in all GWAS
studies to enable cross-study comparisons.</p>
<pre class="r"><code># List all GWAS data files
files &lt;- list.files(
  path = &quot;~/Desktop/autoimmune_gwas&quot;,
  pattern = &quot;\\.h\\.tsv$&quot;,  # Regex to match &quot;.h.tsv&quot; at the end
  full.names = TRUE         # Return full paths
)

# Initialize with variants from first file
#common_variants &lt;- fread(files[1], select = &quot;variant_id&quot;)[[1]]

# Initialize lists to store results
results_list &lt;- list()
duplicate_report &lt;- list()
significant_variants_list &lt;- list()

for (file in files[c(-2, -10, -13, -14)]) {
  # Read relevant columns
  gwas_data &lt;- fread(file, select = c(&quot;variant_id&quot;, &quot;beta&quot;, &quot;standard_error&quot;))
  
  # Filter to common variants and compute Z-scores
  filtered_data &lt;- gwas_data %&gt;% 
    #filter(variant_id %in% common_variants) %&gt;%
    mutate(
      source = gsub(&quot;/Users/zqi/Desktop/autoimmune_gwas/|\\.h\\.tsv$&quot;, &quot;&quot;, file),
      z_hat = beta / standard_error
    )
  
  # Identify and report duplicates
  dup_variants &lt;- filtered_data$variant_id[duplicated(filtered_data$variant_id)]
  if(length(dup_variants) &gt; 0) {
    duplicate_report[[file]] &lt;- data.table(
      file = file,
      variant_id = unique(dup_variants),
      n_duplicates = table(filtered_data$variant_id)[unique(dup_variants)]
    )
  }
  
  # Keep only first occurrence of each variant
  results_list[[file]] &lt;- filtered_data[!duplicated(filtered_data$variant_id), ]
  
  # Select significant variants (|Z-score| &gt; 1.96 for p &lt; 0.05)
  significant_variants &lt;- filtered_data %&gt;%
    filter(abs(z_hat) &gt; 0.5) %&gt;%  # Change threshold as needed
    select(variant_id, beta, standard_error, z_hat) %&gt;%
    mutate(source = gsub(&quot;/Users/zqi/Desktop/autoimmune_gwas/|\\.h\\.tsv$&quot;, &quot;&quot;, file))
  
  significant_variants_list[[file]] &lt;- significant_variants
}

# Combine all studies
combined_results &lt;- bind_rows(results_list)

# Combine all significant variants
significant_variants_df &lt;- bind_rows(significant_variants_list)</code></pre>
<p>Final number of overlapping variants are over 18000. Two traits were
dropped because the overlapping of these two traits will be too low.</p>
</div>
<div id="map-study-ids-to-disease-names" class="section level3">
<h3>Map Study IDs to Disease Names</h3>
<p>Then, replace study identifiers with meaningful disease names for
clearer interpretation.</p>
<pre class="r"><code># Create study-to-disease mapping
study_to_disease &lt;- data.frame(
  source = c(&quot;20190752-GCST000612-EFO_0001060&quot;,
             &quot;23749187-GCST005529-EFO_0003898&quot;,
             &quot;26192919-GCST003044-EFO_0000384&quot;,
             &quot;27329760-GCST003724-EFO_0000289&quot;,
             &quot;23143594-GCST005527-EFO_0000676&quot;,
             &quot;24076602-GCST005531-EFO_0003885&quot;,
             &quot;26192919-GCST003045-EFO_0000729&quot;,
             &quot;33830302-GCST90000529-EFO_0001359&quot;,
             &quot;23143596-GCST005569-EFO_0000685&quot;,
             &quot;26192919-GCST003043-EFO_0003767&quot;,
             &quot;26502338-GCST003156-EFO_0002690&quot;),
  disease = c(&quot;celiac_disease&quot;,
              &quot;ankylosing_spondylitis&quot;,
              &quot;crohns_disease&quot;,
              &quot;biopolar_disorder&quot;,
              &quot;psoriasis&quot;,
              &quot;multiple_sclerosis&quot;,
              &quot;ulcerative_colitis&quot;,
              &quot;type_1_diabetes&quot;,
              &quot;rheumatoid_arthritis&quot;,
              &quot;inflammatory_bowel_disease&quot;,
              &quot;systemic_lupus_erythematosus&quot;)
)

# Merge mapping with results
significant_variants_df &lt;- merge(significant_variants_df, study_to_disease, by = &quot;source&quot;, all.x = TRUE)

# Save processed data
save(significant_variants_df, file = &#39;../data/autoimmune_significant_variants.RData&#39;)</code></pre>
</div>
<div id="create-analysis-ready-matrices" class="section level3">
<h3>Create Analysis-Ready Matrices</h3>
<p>This step is to reshape data into variant-by-disease matrices for
downstream analysis.</p>
<pre class="r"><code>load(&#39;../data/autoimmune_significant_variants.RData&#39;)

# significant_variants_df &lt;- combined_results %&gt;% 
#   filter(abs(z_hat) &gt; 0)</code></pre>
<pre class="r"><code>significant_variants_dedup &lt;- significant_variants_df |&gt;
  distinct(variant_id, disease, .keep_all = TRUE)  # Keeps first duplicate
# Create variant-by-disease beta matrix
beta_matrix &lt;- significant_variants_df %&gt;%
  pivot_wider(
    id_cols = variant_id,
    names_from = disease,
    values_from = beta,
    values_fill = NA
  ) %&gt;% 
  as.data.frame() %&gt;%
  column_to_rownames(&quot;variant_id&quot;) 

z_matrix &lt;- significant_variants_df %&gt;%
  pivot_wider(
    id_cols = variant_id,
    names_from = disease,  # Changed from &#39;disease&#39; to &#39;source&#39; to match your previous code
    values_from = z_hat,
    values_fill = NA  # This explicitly fills missing combinations with NA
  ) %&gt;% 
  as.data.frame() %&gt;%
  column_to_rownames(&quot;variant_id&quot;)

# Save in multiple formats
#save(beta_matrix, file = &quot;beta_matrix.RData&quot;)
#fwrite(beta_matrix, row.names = TRUE, file = &quot;beta_matrix.csv&quot;)</code></pre>
</div>
<div id="semi-nonnegative-matrix-factorization-semi-nmf"
class="section level2">
<h2>Semi-Nonnegative Matrix Factorization (semi-NMF)</h2>
<p>Factorize the effect size matrix into latent factors using semi-NMF,
which allows for capturing only same direction of the disease
mechanism.</p>
<pre class="r"><code>load(&#39;../data/z_matrix.RData&#39;)
Y &lt;- as.matrix(z_matrix)</code></pre>
<pre class="r"><code># Calculate trait-specific variances
var_per_trait &lt;- apply(Y, 2, var, na.rm = TRUE)
s1 &lt;- 0.1 * min(var_per_trait, na.rm = TRUE)  # Set minimum variance threshold

# Initialize timing
timings &lt;- list()

# Run semi-NMF
t0 &lt;- proc.time()
fl0 &lt;- flash(Y,
             ebnm_fn = c(ebnm_point_laplace, ebnm_point_exponential),
             var_type = 0,
             greedy_Kmax = 40,
             nullcheck = FALSE,
             backfit = FALSE,
             verbose = 3)

# Refine with proper variance structure
fl_snmf &lt;- flash_init(Y, var_type = 2, S = s1)
fl_snmf &lt;- flash_factors_init(fl_snmf, fl0, 
                            ebnm_fn = c(ebnm_point_laplace, ebnm_point_exponential))
fl_snmf &lt;- flash_backfit(fl_snmf, extrapolate = FALSE, maxiter = 100, verbose = 3)
fl_snmf &lt;- flash_backfit(fl_snmf, extrapolate = TRUE, maxiter = 100, verbose = 3)
t1 &lt;- proc.time()

# Record timing
timings$fl_snmf &lt;- t1 - t0
print(timings$fl_snmf)

# user     system  elapsed 
# 32.144   2.479  35.119 </code></pre>
<div id="matrix-factorization-with-laplace-priors"
class="section level3">
<h3>Matrix Factorization with Laplace Priors</h3>
<p>Decompose the matrix using Laplace priors for both loadings and
factors, which encourages capturing both positive and negative direction
of the disease mechanism.</p>
<pre class="r"><code>t0 &lt;- proc.time()
fl0_mf &lt;- flash(Y,
                ebnm_fn = c(ebnm_point_laplace, ebnm_point_laplace),
                var_type = 0,
                greedy_Kmax = 40,
                nullcheck = FALSE,
                backfit = FALSE,
                verbose = 3)

# Refine with proper variance structure
fl_mf &lt;- flash_init(Y, var_type = 2, S = s1)
fl_mf &lt;- flash_factors_init(fl_mf, fl0_mf, 
                          ebnm_fn = c(ebnm_point_laplace, ebnm_point_laplace))
fl_mf &lt;- flash_backfit(fl_mf, extrapolate = FALSE, maxiter = 100, verbose = 3)
fl_mf &lt;- flash_backfit(fl_mf, extrapolate = TRUE, maxiter = 100, verbose = 3)
t1 &lt;- proc.time()

# Record timing
timings$fl_mf &lt;- t1 - t0
print(timings$fl_mf)

# user    system  elapsed 
# 131.145   9.444 142.150 </code></pre>
<p>Save the factor loadings for downstream analysis.</p>
<pre class="r"><code># Extract and save factor loadings
fl_mf_ldf &lt;- ldf(fl_mf, type = &quot;i&quot;)
fl_snmf_ldf &lt;- ldf(fl_snmf, type = &quot;i&quot;)

save(fl_snmf_ldf, file = &quot;../output/immune_snmf_factors_sig.RData&quot;)
save(fl_mf_ldf, file = &quot;../output/immune_mf_factors_sig.RData&quot;)</code></pre>
</div>
<div id="explore-and-visualize-factors" class="section level3">
<h3>Explore and Visualize Factors</h3>
<pre class="r"><code>load(&quot;../output/immune_snmf_factors_sig.RData&quot;)
L &lt;- fl_snmf_ldf$L
colnames(L) &lt;- c(paste0(&#39;k&#39;, 1:ncol(L)))
L_df &lt;- as.data.frame(L) %&gt;%
  rownames_to_column(&quot;variant_id&quot;) %&gt;%
  pivot_longer(cols = -variant_id, 
               names_to = &quot;factor&quot;, 
               values_to = &quot;loading&quot;)
# Get top variants per factor
top_variants_snmf &lt;- L_df %&gt;%
  group_by(factor) %&gt;%
  arrange(desc(abs(loading))) %&gt;%
  slice_head(n = 1) %&gt;%  # Top 5 variants per factor
  ungroup()


load(&quot;../output/immune_mf_factors_sig.RData&quot;)
L &lt;- fl_mf_ldf$L
colnames(L) &lt;- c(paste0(&#39;k&#39;, 1:ncol(L)))
L_df &lt;- as.data.frame(L) %&gt;%
  rownames_to_column(&quot;variant_id&quot;) %&gt;%
  pivot_longer(cols = -variant_id, 
               names_to = &quot;factor&quot;, 
               values_to = &quot;loading&quot;)
top_variants_mf &lt;- L_df %&gt;%
  group_by(factor) %&gt;%
  arrange(desc(abs(loading))) %&gt;%
  slice_head(n = 1) %&gt;%  # Top 5 variants per factor
  ungroup()</code></pre>
<div id="heatmap-visualization-of-factors" class="section level4">
<h4>Heatmap Visualization of Factors</h4>
<p>Visualization showing factor effects across traits by heatmap.</p>
<pre class="r"><code>source(&#39;../code/plot_factors.R&#39;)

# Plot semi-NMF factors
F_snmf &lt;- fl_snmf_ldf$F
p_snmf &lt;- plot_factors(F_snmf, row_names = rownames(F_snmf),
                      col_names = paste0(&quot;k&quot;, 1:ncol(F_snmf))) +
          scale_fill_gradient2(low = viridis(3)[1], 
                              high = viridis(3)[2], 
                              name = &quot;Factor Effect&quot;) +
          xlab(&quot;Factor&quot;) +
          ggtitle(&quot;Semi-NMF Factors&quot;) +
          scale_x_discrete(position = &quot;top&quot;) +
          theme(axis.text.x = element_text(angle = 0),
                panel.background = element_rect(fill = &quot;white&quot;),
                axis.ticks = element_blank())
# Scale for fill is already present.
# Adding another scale for fill, which will replace the existing scale.

# Plot Laplace MF factors
F_mf &lt;- fl_mf_ldf$F
p_mf &lt;- plot_factors(F_mf, row_names = rownames(F_mf),
                    col_names = paste0(&quot;k&quot;, 1:ncol(F_mf))) +
        scale_fill_gradient2(low = viridis(3)[1], 
                            high = viridis(3)[2], 
                            name = &quot;Factor Effect&quot;) +
        xlab(&quot;Factors&quot;) +
        ggtitle(&quot;Laplace MF Factors&quot;) +
        scale_x_discrete(position = &quot;top&quot;) +
        theme(axis.text.x = element_text(angle = 0),
              panel.background = element_rect(fill = &quot;white&quot;),
              axis.ticks = element_blank())
# Scale for fill is already present.
# Adding another scale for fill, which will replace the existing scale.

# Combine plots
plot_grid(p_snmf, p_mf, nrow = 2, ncol = 1)</code></pre>
<p><img src="figure/autoimmune_factors_significant.Rmd/factor-heatmaps-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-factor-heatmaps-1">
Past versions of factor-heatmaps-1.png
</button>
</p>
<div id="fig-factor-heatmaps-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/9fba759b8c919eaedca7f9ea1af24a1da0a41091/docs/figure/autoimmune_factors_significant.Rmd/factor-heatmaps-1.png" target="_blank">9fba759</a>
</td>
<td>
zq2209
</td>
<td>
2025-06-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
</div>
<div id="visualization-of-loadings" class="section level3">
<h3>Visualization of Loadings</h3>
<p>Extract high loading variants for each loadings</p>
<pre class="r"><code>L &lt;- fl_mf_ldf$L
colnames(L) &lt;- c(paste0(&#39;k&#39;, 1:ncol(L)))

L_df &lt;- as.data.frame(L) %&gt;%
  rownames_to_column(&quot;variant_id&quot;) %&gt;%
  pivot_longer(cols = -variant_id, 
               names_to = &quot;factor&quot;, 
               values_to = &quot;loading&quot;)

merged_data &lt;- combined_results %&gt;%
  inner_join(L_df, by = &quot;variant_id&quot;)



# Get top 5 variants per factor
top_variants &lt;- L_df %&gt;%
  group_by(factor) %&gt;%
  arrange(desc(abs(loading))) %&gt;%
  slice(1:5) %&gt;%  # Top 5 variants per factor
  ungroup()

# Merge with effect data
plot_data &lt;- combined_results %&gt;%
  inner_join(top_variants, by = &quot;variant_id&quot;) %&gt;%
  mutate(
    ci_low = beta - 1.96 * standard_error,
    ci_high = beta + 1.96 * standard_error,
    # Label for the variant (shows loading value)
    variant_label = paste0(variant_id, &quot; (λ=&quot;, round(loading, 2), &quot;)&quot;)
  )</code></pre>
<pre class="r"><code>plot_single_variant &lt;- function(variant_data) {
  # Prepare labels
  label_text &lt;- cbind(
    c(&quot;Trait&quot;, variant_data$disease),
    c(&quot;Effect [95% CI]&quot;, 
      sprintf(&quot;%.2f [%.2f, %.2f]&quot;, 
              variant_data$beta, 
              variant_data$ci_low, 
              variant_data$ci_high))
  )
  
  forestplot(
    labeltext = label_text,
    mean = c(NA, variant_data$beta),
    lower = c(NA, variant_data$ci_low),
    upper = c(NA, variant_data$ci_high),
    title = paste(&quot;Variant:&quot;, variant_data$variant_label[1], 
                 &quot;\nFactor:&quot;, variant_data$factor[1]),
    xticks = seq(round(min(variant_data$ci_low))-0.1, round(max(variant_data$ci_high))+0.1, 0.05),
    boxsize = 0.2,
    lineheight = unit(0.8, &quot;cm&quot;),
    col = fpColors(box = &quot;royalblue&quot;, line = &quot;darkblue&quot;, summary = &quot;royalblue&quot;),
    graphwidth = unit(6, &quot;cm&quot;),
    clip = c(-1, 1)  # Force consistent x-axis limits
  )
}

nested_plots &lt;- plot_data %&gt;%
  group_by(factor, variant_id) %&gt;%
  group_split()

plot_single_variant(nested_plots[[1]])

all_plots &lt;- map(nested_plots, plot_single_variant)

walk2(all_plots, nested_plots, ~{
  factor_dir &lt;- file.path(&quot;../output/forest_plots/significant_signals/mf/&quot;, unique(.y$factor))
  if (!dir.exists(factor_dir)) dir.create(factor_dir, recursive = TRUE)
  
  png(file.path(factor_dir, paste0(unique(.y$variant_id), &quot;.png&quot;)),
      width = 1200, height = 800, res = 150)
  print(.x)
  dev.off()
})</code></pre>
<pre class="r"><code># Get path of all png files
snmf_png_files &lt;- file.path(&quot;..&quot;, &quot;output&quot;, &quot;forest_plots&quot;, &quot;significant_signals&quot;, &quot;snmf&quot;, 
                         top_variants_snmf$factor, 
                         paste0(top_variants_snmf$variant_id, &quot;.png&quot;))

# Read images into a list
# Load PNGs as raster objects
plots &lt;- lapply(snmf_png_files, function(x) {
  img &lt;- readPNG(x)
  rasterGrob(img, interpolate = TRUE)
})

# Arrange in a grid (e.g., 4x2)
grid.arrange(grobs = plots, ncol = 2)</code></pre>
<p><img src="figure/autoimmune_factors_significant.Rmd/unnamed-chunk-3-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-3-1">
Past versions of unnamed-chunk-3-1.png
</button>
</p>
<div id="fig-unnamed-chunk-3-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/9fba759b8c919eaedca7f9ea1af24a1da0a41091/docs/figure/autoimmune_factors_significant.Rmd/unnamed-chunk-3-1.png" target="_blank">9fba759</a>
</td>
<td>
zq2209
</td>
<td>
2025-06-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code># Get path of all png files
mf_png_files &lt;- file.path(&quot;..&quot;, &quot;output&quot;, &quot;forest_plots&quot;, &quot;significant_signals&quot;, &quot;mf&quot;, 
                         top_variants_mf$factor, 
                         paste0(top_variants_mf$variant_id, &quot;.png&quot;))

# Read images into a list
# Load PNGs as raster objects
plots &lt;- lapply(mf_png_files, function(x) {
  img &lt;- readPNG(x)
  rasterGrob(img, interpolate = TRUE)
})

# Arrange in a grid (e.g., 4x2)
grid.arrange(grobs = plots, ncol = 2)</code></pre>
<p><img src="figure/autoimmune_factors_significant.Rmd/unnamed-chunk-4-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-unnamed-chunk-4-1">
Past versions of unnamed-chunk-4-1.png
</button>
</p>
<div id="fig-unnamed-chunk-4-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/9fba759b8c919eaedca7f9ea1af24a1da0a41091/docs/figure/autoimmune_factors_significant.Rmd/unnamed-chunk-4-1.png" target="_blank">9fba759</a>
</td>
<td>
zq2209
</td>
<td>
2025-06-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.4.3 (2025-02-28)
# Platform: aarch64-apple-darwin20
# Running under: macOS Sequoia 15.5
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] grid      stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
#  [1] fastTopics_0.6-192 flashier_1.0.56    ebnm_1.1-34        viridis_0.6.5     
#  [5] viridisLite_0.4.2  cowplot_1.1.3      reshape2_1.4.4     gridExtra_2.3     
#  [9] png_0.1-8          forestplot_3.1.6   abind_1.4-8        checkmate_2.3.2   
# [13] pheatmap_1.0.12    data.table_1.17.4  lubridate_1.9.4    forcats_1.0.0     
# [17] stringr_1.5.1      dplyr_1.1.4        purrr_1.0.4        readr_2.1.5       
# [21] tidyr_1.3.1        tibble_3.2.1       ggplot2_3.5.2      tidyverse_2.0.0   
# 
# loaded via a namespace (and not attached):
#  [1] tidyselect_1.2.1     farver_2.1.2         lazyeval_0.2.2      
#  [4] fastmap_1.2.0        promises_1.3.3       digest_0.6.37       
#  [7] timechange_0.3.0     lifecycle_1.0.4      invgamma_1.1        
# [10] magrittr_2.0.3       compiler_4.4.3       progress_1.2.3      
# [13] rlang_1.1.6          sass_0.4.10          tools_4.4.3         
# [16] yaml_2.3.10          knitr_1.50           labeling_0.4.3      
# [19] prettyunits_1.2.0    htmlwidgets_1.6.4    scatterplot3d_0.3-44
# [22] plyr_1.8.9           RColorBrewer_1.1-3   Rtsne_0.17          
# [25] workflowr_1.7.1      withr_3.0.2          git2r_0.36.2        
# [28] colorspace_2.1-1     scales_1.4.0         gtools_3.9.5        
# [31] cli_3.6.5            crayon_1.5.3         rmarkdown_2.29      
# [34] generics_0.1.4       RcppParallel_5.1.10  rstudioapi_0.17.1   
# [37] httr_1.4.7           tzdb_0.5.0           pbapply_1.7-2       
# [40] cachem_1.1.0         splines_4.4.3        parallel_4.4.3      
# [43] softImpute_1.4-3     vctrs_0.6.5          Matrix_1.7-3        
# [46] jsonlite_2.0.0       hms_1.1.3            mixsqp_0.3-54       
# [49] ggrepel_0.9.6        irlba_2.3.5.1        horseshoe_0.2.0     
# [52] trust_0.1-8          plotly_4.10.4        jquerylib_0.1.4     
# [55] glue_1.8.0           uwot_0.2.3           stringi_1.8.7       
# [58] Polychrome_1.5.4     gtable_0.3.6         later_1.4.2         
# [61] quadprog_1.5-8       pillar_1.10.2        htmltools_0.5.8.1   
# [64] truncnorm_1.0-9      R6_2.6.1             rprojroot_2.0.4     
# [67] evaluate_1.0.3       lattice_0.22-7       backports_1.5.0     
# [70] RhpcBLASctl_0.23-42  SQUAREM_2021.1       ashr_2.2-63         
# [73] httpuv_1.6.16        bslib_0.9.0          Rcpp_1.0.14         
# [76] deconvolveR_1.2-1    whisker_0.4.1        xfun_0.52           
# [79] fs_1.6.6             pkgconfig_2.0.3</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
