<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Zining Qi" />


<title>Factor Analysis of Autoimmune GWAS Data using flashier</title>

<script src="site_libs/header-attrs-2.29/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">ebmf_immune</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/zq2209/ebmf_immune">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Factor Analysis of Autoimmune GWAS Data
using flashier</h1>
<h4 class="author">Zining Qi</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-05-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>ebmf_immune/analysis/</code>
<span class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity itâ€™s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree1f8e8279fa757011449326e83d353ea445c7f024targetblank1f8e827a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/zq2209/ebmf_immune/tree/1f8e8279fa757011449326e83d353ea445c7f024" target="_blank">1f8e827</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomzq2209ebmfimmunetree1f8e8279fa757011449326e83d353ea445c7f024targetblank1f8e827a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/zq2209/ebmf_immune/tree/1f8e8279fa757011449326e83d353ea445c7f024" target="_blank">1f8e827</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .Rproj.user/

</code></pre>
<p>
Note that any generated files, e.g.Â HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">
<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/autoimmune_factors.Rmd</code>)
and HTML (<code>docs/autoimmune_factors.html</code>) files. If youâ€™ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/1f8e8279fa757011449326e83d353ea445c7f024/analysis/autoimmune_factors.Rmd" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/zq2209/ebmf_immune/1f8e8279fa757011449326e83d353ea445c7f024/docs/autoimmune_factors.html" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/30a211dc8958e5c48fc86c3e53831a62ed27ac07/analysis/autoimmune_factors.Rmd" target="_blank">30a211d</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/ef8b718330f0543c05d457c3f7fbc2cf94e64892/analysis/autoimmune_factors.Rmd" target="_blank">ef8b718</a>
</td>
<td>
Zining Qi
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/865d316b479332a8baac65333fa6e196f03dd2bd/analysis/autoimmune_factors.Rmd" target="_blank">865d316</a>
</td>
<td>
Zining Qi
</td>
<td>
2025-05-01
</td>
<td>
update
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This analysis applies Empirical Bayes Matrix Factorization methods to
genome-wide association study (GWAS) data for multiple autoimmune
diseases. We use the <code>flashier</code> package to:</p>
<ol style="list-style-type: decimal">
<li>Perform semi-nonnegative matrix factorization (semi-NMF)</li>
<li>Perform matrix factorization with Laplace priors for both L and
F</li>
<li>Compare and visualize the resulting factors</li>
<li>Explore factor-trait relationships by affected biological
systems</li>
</ol>
<div id="setup" class="section level3">
<h3>Setup</h3>
<p>Load required packages and data:</p>
<pre class="r"><code># Data manipulation and visualization
library(tidyverse)
# â”€â”€ Attaching core tidyverse packages â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse 2.0.0 â”€â”€
# âœ” dplyr     1.1.4     âœ” readr     2.1.5
# âœ” forcats   1.0.0     âœ” stringr   1.5.1
# âœ” ggplot2   3.5.2     âœ” tibble    3.2.1
# âœ” lubridate 1.9.4     âœ” tidyr     1.3.1
# âœ” purrr     1.0.4     
# â”€â”€ Conflicts â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€â”€ tidyverse_conflicts() â”€â”€
# âœ– dplyr::filter() masks stats::filter()
# âœ– dplyr::lag()    masks stats::lag()
# â„¹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors
library(reshape2)
# 
# Attaching package: &#39;reshape2&#39;
# 
# The following object is masked from &#39;package:tidyr&#39;:
# 
#     smiths
library(cowplot)
# 
# Attaching package: &#39;cowplot&#39;
# 
# The following object is masked from &#39;package:lubridate&#39;:
# 
#     stamp
library(viridis)
# Loading required package: viridisLite

# Matrix factorization tools
library(flashier)
# Loading required package: ebnm
library(fastTopics)

# Load preprocessed GWAS effect size matrix
load(&quot;../data/beta_matrix.RData&quot;)</code></pre>
</div>
<div id="data-overview" class="section level3">
<h3>Data Overview</h3>
<p>The input data is a matrix of GWAS effect sizes (<span
class="math inline">\(\beta\)</span>) where:</p>
<ul>
<li>Rows represent genetic variants</li>
<li>Columns represent autoimmune diseases</li>
</ul>
<pre class="r"><code># Preview the data structure
head(beta_matrix[, 1:5])

# Convert to matrix format for flashier
Y &lt;- as.matrix(beta_matrix)
#            celiac_disease   psoriasis rheumatoid_arthritis
# rs3125037      0.01980263  0.03912455         0.0005998201
# rs3922851     -0.05413945 -0.01582455        -0.0038072383
# rs11252923     0.01783992 -0.09167679        -0.0259333820
# rs4881399      0.02761517 -0.09255398        -0.0236781274
# rs1044261      0.01685712 -0.07601749         0.0042907814
# rs7075141     -0.06421857 -0.01877516        -0.0025031302
#            ankylosing_spondylitis multiple_sclerosis
# rs3125037            0.0007996802        0.026641931
# rs3922851           -0.0081329832        0.010939940
# rs11252923          -0.0012007206        0.001998003
# rs4881399           -0.0022024236        0.001998003
# rs1044261            0.0115332358        0.019802627
# rs7075141           -0.0023026491       -0.023678127</code></pre>
</div>
<div id="semi-nonnegative-matrix-factorization-semi-nmf"
class="section level2">
<h2>Semi-Nonnegative Matrix Factorization (semi-NMF)</h2>
<p>Factorize the effect size matrix into latent factors using semi-NMF,
which allows for capturing only same direction of the disease
mechanism.</p>
<pre class="r"><code># Calculate trait-specific variances
var_per_trait &lt;- apply(beta_matrix, 2, var, na.rm = TRUE)
s1 &lt;- 0.1 * min(var_per_trait, na.rm = TRUE)  # Set minimum variance threshold

# Initialize timing
timings &lt;- list()

# Run semi-NMF
t0 &lt;- proc.time()
fl0 &lt;- flash(Y,
             ebnm_fn = c(ebnm_point_laplace, ebnm_point_exponential),
             var_type = 0,
             greedy_Kmax = 40,
             nullcheck = FALSE,
             backfit = FALSE,
             verbose = 3)

# Refine with proper variance structure
fl_snmf &lt;- flash_init(Y, var_type = 2, S = s1)
fl_snmf &lt;- flash_factors_init(fl_snmf, fl0, 
                            ebnm_fn = c(ebnm_point_laplace, ebnm_point_exponential))
fl_snmf &lt;- flash_backfit(fl_snmf, extrapolate = FALSE, maxiter = 100, verbose = 3)
fl_snmf &lt;- flash_backfit(fl_snmf, extrapolate = TRUE, maxiter = 100, verbose = 3)
t1 &lt;- proc.time()

# Record timing
timings$fl_snmf &lt;- t1 - t0
print(timings$fl_snmf)

# user     system  elapsed 
# 99.693   8.619   110.916 </code></pre>
<p>user system elapsed 99.693 8.619 110.916</p>
<div id="matrix-factorization-with-laplace-priors"
class="section level3">
<h3>Matrix Factorization with Laplace Priors</h3>
<p>Decompose the matrix using Laplace priors for both loadings and
factors, which encourages capturing both positive and negative direction
of the disease mechanism.</p>
<pre class="r"><code>t0 &lt;- proc.time()
fl0_mf &lt;- flash(Y,
                ebnm_fn = c(ebnm_point_laplace, ebnm_point_laplace),
                var_type = 0,
                greedy_Kmax = 40,
                nullcheck = FALSE,
                backfit = FALSE,
                verbose = 3)

# Refine with proper variance structure
fl_mf &lt;- flash_init(Y, var_type = 2, S = s1)
fl_mf &lt;- flash_factors_init(fl_mf, fl0_mf, 
                          ebnm_fn = c(ebnm_point_laplace, ebnm_point_laplace))
fl_mf &lt;- flash_backfit(fl_mf, extrapolate = FALSE, maxiter = 100, verbose = 3)
fl_mf &lt;- flash_backfit(fl_mf, extrapolate = TRUE, maxiter = 100, verbose = 3)
t1 &lt;- proc.time()

# Record timing
timings$fl_mf &lt;- t1 - t0
print(timings$fl_mf)

# user    system  elapsed 
# 108.067 7.943   117.655 </code></pre>
<p>Save the factor loadings for downstream analysis.</p>
<pre class="r"><code># Extract and save factor loadings
fl_mf_ldf &lt;- ldf(fl_mf, type = &quot;i&quot;)
fl_snmf_ldf &lt;- ldf(fl_snmf, type = &quot;i&quot;)

save(fl_snmf_ldf, file = &quot;output/immune_snmf_factors.RData&quot;)
save(fl_mf_ldf, file = &quot;output/immune_mf_factors.RData&quot;)</code></pre>
</div>
<div id="explore-and-visualize-factors" class="section level3">
<h3>Explore and Visualize Factors</h3>
<pre class="r"><code>load(&quot;../output/immune_mf_factors.RData&quot;)
load(&quot;../output/immune_snmf_factors.RData&quot;)</code></pre>
<div id="structure-plots" class="section level4">
<h4>Structure Plots</h4>
<p>Visualize the factor patterns across traits for both methods by using
strcture plot.</p>
<pre class="r"><code># Define trait order for consistent visualization
traits &lt;- colnames(Y)
traits &lt;- factor(traits,
                 levels = c(&quot;celiac_disease&quot;, &quot;psoriasis&quot;, &quot;rheumatoid_arthritis&quot;,
                            &quot;ankylosing_spondylitis&quot;, &quot;multiple_sclerosis&quot;, 
                            &quot;inflammatory_bowel_disease&quot;, &quot;crohns_disease&quot;, 
                            &quot;ulcerative_colitis&quot;, &quot;systemic_lupus_erythematosus&quot;,
                            &quot;biopolar_disorder&quot;, &quot;type_1_diabetes&quot;))
trait_factors &lt;- c(1:8)

# Source plotting functions
source(&#39;../code/sturcture_plot.R&#39;)
# Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
# â„¹ Please use tidy evaluation idioms with `aes()`.
# â„¹ See also `vignette(&quot;ggplot2-in-packages&quot;)` for more information.
# This warning is displayed once every 8 hours.
# Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
# generated.

# Create comparison plot
plot_grid(p1, p2, nrow = 2, ncol = 1)</code></pre>
<p><img src="figure/autoimmune_factors.Rmd/structure-plots-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-1">
Past versions of structure-plots-1.png
</button>
</p>
<div id="fig-structure-plots-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/1f8e8279fa757011449326e83d353ea445c7f024/docs/figure/autoimmune_factors.Rmd/structure-plots-1.png" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="heatmap-visualization-of-factors" class="section level4">
<h4>Heatmap Visualization of Factors</h4>
<p>Alternative visualization showing factor effects across traits by
heatmap.</p>
<pre class="r"><code>source(&#39;../code/plot_factors.R&#39;)

# Plot semi-NMF factors
F_snmf &lt;- fl_snmf_ldf$F
p_snmf &lt;- plot_factors(F_snmf, row_names = rownames(F_snmf),
                      col_names = paste0(&quot;k&quot;, 1:ncol(F_snmf))) +
          scale_fill_gradient2(low = viridis(3)[1], 
                              high = viridis(3)[2], 
                              name = &quot;Factor Effect&quot;) +
          xlab(&quot;Factor&quot;) +
          ggtitle(&quot;Semi-NMF Factors&quot;) +
          scale_x_discrete(position = &quot;top&quot;) +
          theme(axis.text.x = element_text(angle = 0),
                panel.background = element_rect(fill = &quot;white&quot;),
                axis.ticks = element_blank())
# Scale for fill is already present.
# Adding another scale for fill, which will replace the existing scale.

# Plot Laplace MF factors
F_mf &lt;- fl_mf_ldf$F
p_mf &lt;- plot_factors(F_mf, row_names = rownames(F_mf),
                    col_names = paste0(&quot;k&quot;, 1:ncol(F_mf))) +
        scale_fill_gradient2(low = viridis(3)[1], 
                            high = viridis(3)[2], 
                            name = &quot;Factor Effect&quot;) +
        xlab(&quot;Factors&quot;) +
        ggtitle(&quot;Laplace MF Factors&quot;) +
        scale_x_discrete(position = &quot;top&quot;) +
        theme(axis.text.x = element_text(angle = 0),
              panel.background = element_rect(fill = &quot;white&quot;),
              axis.ticks = element_blank())
# Scale for fill is already present.
# Adding another scale for fill, which will replace the existing scale.

# Combine plots
plot_grid(p_snmf, p_mf, nrow = 2, ncol = 1)</code></pre>
<p><img src="figure/autoimmune_factors.Rmd/factor-heatmaps-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-factor-heatmaps-1">
Past versions of factor-heatmaps-1.png
</button>
</p>
<div id="fig-factor-heatmaps-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/1f8e8279fa757011449326e83d353ea445c7f024/docs/figure/autoimmune_factors.Rmd/factor-heatmaps-1.png" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>From both structure plot and heatmap, from semi-NMF model, K2 is
dominantly impact IBD, CD, and UC. K3 have negative impact on IBD, CD,
and UC. K3 have negative impact on IBD, CD, and UC. And positive impact
on celiac disease. K6 is dominant for bipolar disorder, both positive
and negative effects. K2 also have a pretty strong negative impact on
systemic lupus erythematosus.</p>
</div>
</div>
<div id="factor-analysis-by-affected-biological-system"
class="section level3">
<h3>Factor Analysis by Affected Biological System</h3>
<p>Explore whether factors correspond to specific biological systems
affected by the autoimmune diseases.</p>
<pre class="r"><code>F &lt;- fl_snmf_ldf$F
traits &lt;- rownames(F)

affected_system &lt;- c(
  &quot;gastrointestinal&quot;, &quot;Skin&quot;, &quot;joints_muscles&quot;, 
  &quot;spine_sacroiliac_joints&quot;, &quot;nervous&quot;, 
  &quot;gastrointestinal&quot;, &quot;gastrointestinal&quot;, 
  &quot;gastrointestinal&quot;, &quot;multiple&quot;, 
  &quot;nervous&quot;, &quot;endocrine_glands&quot;
)

anno_data &lt;- plot_factor_annotation_data(F, affected_system)

# Create heatmap
p_snmf_anno &lt;- pheatmap::pheatmap(
  anno_data$F_ordered,
  annotation_row = anno_data$annotation_df,
  annotation_colors = anno_data$annotation_colors,
  cluster_cols = FALSE,
  cluster_rows = FALSE,
  scale = &#39;none&#39;,
  show_colnames = TRUE,
  border_color = &quot;white&quot;,
  main = &quot;Semi-NMF Factors by Affected Biological System&quot;,
  color = colorRampPalette(c(&quot;white&quot;, &quot;#21908CFF&quot;))(100)
)

p_snmf_anno</code></pre>
<p><img src="figure/autoimmune_factors.Rmd/affect-system-heatmap-1.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-affect-system-heatmap-1">
Past versions of affect-system-heatmap-1.png
</button>
</p>
<div id="fig-affect-system-heatmap-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/1f8e8279fa757011449326e83d353ea445c7f024/docs/figure/autoimmune_factors.Rmd/affect-system-heatmap-1.png" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<pre class="r"><code>
F &lt;- fl_mf_ldf$F
traits &lt;- rownames(F)

anno_data &lt;- plot_factor_annotation_data(F, affected_system)

# Create heatmap
p_mf_anno &lt;- pheatmap::pheatmap(
  anno_data$F_ordered,
  annotation_row = anno_data$annotation_df,
  annotation_colors = anno_data$annotation_colors,
  cluster_cols = FALSE,
  cluster_rows = FALSE,
  scale = &#39;none&#39;,
  show_colnames = TRUE,
  border_color = &quot;white&quot;,
  main = &quot;MF Factors by Affected System&quot;,
  color = colorRampPalette(c(&quot;#440154FF&quot;, &quot;white&quot;, &quot;#21908CFF&quot;))(100)
)

p_mf_anno</code></pre>
<p><img src="figure/autoimmune_factors.Rmd/affect-system-heatmap-2.png" width="840" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-affect-system-heatmap-2">
Past versions of affect-system-heatmap-2.png
</button>
</p>
<div id="fig-affect-system-heatmap-2" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/zq2209/ebmf_immune/blob/1f8e8279fa757011449326e83d353ea445c7f024/docs/figure/autoimmune_factors.Rmd/affect-system-heatmap-2.png" target="_blank">1f8e827</a>
</td>
<td>
â€œJohn
</td>
<td>
2025-05-01
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>For K2 and K3, they have impact on disease that have the same
affected system, gastrointestinal.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.4.3 (2025-02-28)
# Platform: aarch64-apple-darwin20
# Running under: macOS Sequoia 15.4.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
#  [1] fastTopics_0.6-192 flashier_1.0.56    ebnm_1.1-34        viridis_0.6.5     
#  [5] viridisLite_0.4.2  cowplot_1.1.3      reshape2_1.4.4     lubridate_1.9.4   
#  [9] forcats_1.0.0      stringr_1.5.1      dplyr_1.1.4        purrr_1.0.4       
# [13] readr_2.1.5        tidyr_1.3.1        tibble_3.2.1       ggplot2_3.5.2     
# [17] tidyverse_2.0.0   
# 
# loaded via a namespace (and not attached):
#  [1] tidyselect_1.2.1     farver_2.1.2         fastmap_1.2.0       
#  [4] lazyeval_0.2.2       promises_1.3.2       digest_0.6.37       
#  [7] timechange_0.3.0     lifecycle_1.0.4      invgamma_1.1        
# [10] magrittr_2.0.3       compiler_4.4.3       rlang_1.1.6         
# [13] sass_0.4.10          progress_1.2.3       tools_4.4.3         
# [16] yaml_2.3.10          data.table_1.17.0    knitr_1.50          
# [19] labeling_0.4.3       prettyunits_1.2.0    htmlwidgets_1.6.4   
# [22] scatterplot3d_0.3-44 plyr_1.8.9           RColorBrewer_1.1-3  
# [25] Rtsne_0.17           workflowr_1.7.1      withr_3.0.2         
# [28] grid_4.4.3           git2r_0.36.2         colorspace_2.1-1    
# [31] scales_1.4.0         gtools_3.9.5         cli_3.6.5           
# [34] rmarkdown_2.29       crayon_1.5.3         generics_0.1.3      
# [37] RcppParallel_5.1.10  rstudioapi_0.17.1    httr_1.4.7          
# [40] tzdb_0.5.0           pbapply_1.7-2        cachem_1.1.0        
# [43] splines_4.4.3        parallel_4.4.3       softImpute_1.4-1    
# [46] vctrs_0.6.5          Matrix_1.7-3         jsonlite_2.0.0      
# [49] hms_1.1.3            mixsqp_0.3-54        ggrepel_0.9.6       
# [52] irlba_2.3.5.1        horseshoe_0.2.0      trust_0.1-8         
# [55] plotly_4.10.4        jquerylib_0.1.4      glue_1.8.0          
# [58] uwot_0.2.3           stringi_1.8.7        Polychrome_1.5.4    
# [61] gtable_0.3.6         later_1.4.2          quadprog_1.5-8      
# [64] pillar_1.10.2        htmltools_0.5.8.1    truncnorm_1.0-9     
# [67] R6_2.6.1             rprojroot_2.0.4      evaluate_1.0.3      
# [70] lattice_0.22-6       pheatmap_1.0.12      RhpcBLASctl_0.23-42 
# [73] SQUAREM_2021.1       ashr_2.2-63          httpuv_1.6.15       
# [76] bslib_0.9.0          Rcpp_1.0.14          gridExtra_2.3       
# [79] deconvolveR_1.2-1    whisker_0.4.1        xfun_0.52           
# [82] fs_1.6.6             pkgconfig_2.0.3</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
