<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>Initial exploration of the pancreas data set</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Initial exploration of the pancreas data
set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2024-11-01
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree42a24a5ec6d78983a037e8d0021884412912a85atargetblank42a24a5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/42a24a5ec6d78983a037e8d0021884412912a85a" target="_blank">42a24a5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree42a24a5ec6d78983a037e8d0021884412912a85atargetblank42a24a5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/42a24a5ec6d78983a037e8d0021884412912a85a" target="_blank">42a24a5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/human_pancreas_norm_complexBatch.h5ad
    Untracked:  data/pancreas.RData

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pancreas.Rmd</code>) and HTML
(<code>docs/pancreas.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/42a24a5ec6d78983a037e8d0021884412912a85a/analysis/pancreas.Rmd" target="_blank">42a24a5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-01
</td>
<td>
workflowr::wflow_publish("pancreas.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/4349f73b39837f5fb9cb6a8bfd8cee72ac5dd587/analysis/pancreas.Rmd" target="_blank">4349f73</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
Added some code chunks to run seurat on the pancreas data.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/108b85391487fde70d4c859ce2e33b4bebe6e6ce/docs/pancreas.html" target="_blank">108b853</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
Switched from umap to tsne in the pancreas analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/fca056b6dd4c29da4c05118c6956d2e99a3a4a7f/analysis/pancreas.Rmd" target="_blank">fca056b</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
workflowr::wflow_publish("pancreas.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/833a14700f5838094e0086f2d237026a880103c7/docs/pancreas.html" target="_blank">833a147</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a8ffa99b330febbbfde407c468093ea329924fd5/analysis/pancreas.Rmd" target="_blank">a8ffa99</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
workflowr::wflow_publish("pancreas.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/25f389a7a147366e9fb9f71c43de7664c5e03611/docs/pancreas.html" target="_blank">25f389a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
Added umap plots to pancreas analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/32cb6258d8e13598fddb24886adf78e7a37a2368/analysis/pancreas.Rmd" target="_blank">32cb625</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
workflowr::wflow_publish("pancreas.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a04b2c0aa0d3167f6a0e5d0d727e751371c1f9aa/analysis/pancreas.Rmd" target="_blank">a04b2c0</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
A few improvements to the pancreas workflowr analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/fed31e6dbf128b29037d0085f9e21eb008ea58c8/docs/pancreas.html" target="_blank">fed31e6</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
First build of the pancreas workflowr page.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/886cd01bf5600732824ca58fa7e7bd43a6691ee5/analysis/pancreas.Rmd" target="_blank">886cd01</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
<td>
workflowr::wflow_publish("pancreas.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/528d5efba05e45765dced804df846bd61d3a0a27/docs/pancreas.html" target="_blank">528d5ef</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
First build of the pancreas workflowr page.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a1d7a174ecd729f2c3cb1da33883270b0640eb0b/analysis/pancreas.Rmd" target="_blank">a1d7a17</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
Added some background on the pancreas data set.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3fa28da165fabed976330431645f3cd90a00b395/analysis/pancreas.Rmd" target="_blank">3fa28da</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
Small edit to pancreas.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/370a3360af1ec0dea5cdba677563b60b00056e27/analysis/pancreas.Rmd" target="_blank">370a336</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
Still working on pancreas.Rmd.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/fbe0b62dd432561ebec724b0b85b689d3b60221c/analysis/pancreas.Rmd" target="_blank">fbe0b62</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
Made a few improvements to the pancreas analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/c090d1d9ab24782cb6b9299f976c8939872caaa3/analysis/pancreas.Rmd" target="_blank">c090d1d</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-30
</td>
<td>
Working on initial examination of pancreas data.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The aim of this analysis is to take an initial look at the “pancreas”
data set that was featured in in the <a
href="https://doi.org/10.1038/s41592-021-01336-8">Luecken et al 2022
benchmarking paper</a>, and prepare the data in a convenient form for
subsequent analyses in R.</p>
<p>In addition to being featured in the Luecken et al paper, it has been
used in several papers on “data integration” methods for single-cell
data (also known as “batch correction” or “harmonization” methods). See
for example the <a href="https://doi.org/10.1038/nbt.4091">MNN
paper</a>. The Supplementary Note in the Luecken et al paper has
additional references.</p>
<p>See Supplementary Fig. 13, Supplementary Note 3 and Supplementary
Data 7 of the Luecken et al paper for more information on this data
set.</p>
<p>First, load the packages needed for this analysis. Note that
MatrixExtra is also used in one of the steps below.</p>
<pre class="r"><code>library(tools)
library(Matrix)
library(hdf5r)
library(rsvd)
library(Rtsne)
library(ggplot2)
library(cowplot)
library(Seurat)</code></pre>
<p>Download the file “human_pancreas_norm_complexBatch.h5ad” from <a
href=""
title="https://figshare.com/articles/dataset/Benchmarking_atlas-level_data_integration_in_single-cell_genomics_-_integration_task_datasets_Immune_and_pancreas_/12420968">figshare</a>
and copy it to the “data” subdirectory of this git repository. Then load
the count data, and encode them as a sparse matrix:</p>
<pre class="r"><code>dat &lt;- H5File$new(&quot;../data/human_pancreas_norm_complexBatch.h5ad&quot;,mode = &quot;r&quot;)
counts &lt;- dat[[&quot;layers&quot;]][[&quot;counts&quot;]][,]
counts &lt;- t(counts)
counts &lt;- as(counts,&quot;CsparseMatrix&quot;)
sample_info &lt;- data.frame(id          = dat[[&quot;obs&quot;]][[&quot;_index&quot;]][],
                          tech        = dat[[&quot;obs&quot;]][[&quot;tech&quot;]][],
                          celltype    = dat[[&quot;obs&quot;]][[&quot;celltype&quot;]][],
                          size_factor = dat[[&quot;obs&quot;]][[&quot;size_factors&quot;]][],
                          stringsAsFactors = FALSE)
sample_info &lt;- transform(sample_info,
                         tech     = factor(tech),
                         celltype = factor(celltype))
levels(sample_info$tech)     &lt;- dat[[&quot;obs&quot;]][[&quot;__categories&quot;]][[&quot;tech&quot;]][]
levels(sample_info$celltype) &lt;- dat[[&quot;obs&quot;]][[&quot;__categories&quot;]][[&quot;celltype&quot;]][]
genes &lt;- dat[[&quot;var&quot;]][[&quot;_index&quot;]][]
rownames(counts) &lt;- sample_info$id
colnames(counts) &lt;- genes</code></pre>
<p>Note that some of the data are not actually counts, so perhaps
calling this matrix “counts” is a bit misleading. Regardless, in some of
our analyses we will model these data as counts.</p>
<p>Also note that in Luecken et al the counts were log-transformed, but
here we taking the untransformed data.</p>
<p>The matrix has 16,382 rows (cells) and 19,093 columns (genes), and
about 18% of the entries are nonzeros:</p>
<pre class="r"><code>nrow(counts)
ncol(counts)
mean(counts &gt; 0)
# [1] 16382
# [1] 19093
# [1] 0.1779012</code></pre>
<p>The pancreas data are actually a combination of several scRNA-seq
data sets that are from different sequencing technologies or were
processed in different ways:</p>
<pre class="r"><code>table(sample_info$tech)
# 
#     celseq    celseq2 fluidigmc1    inDrop1    inDrop2    inDrop3    inDrop4 
#       1004       2285        638       1937       1724       3605       1303 
#    smarter  smartseq2 
#       1492       2394</code></pre>
<p>The cells were previously annotated by cell type:</p>
<pre class="r"><code>table(sample_info$celltype)
# 
#             acinar activated_stellate              alpha               beta 
#               1669                464               5493               4169 
#              delta             ductal        endothelial            epsilon 
#               1055               2142                313                 32 
#              gamma         macrophage               mast quiescent_stellate 
#                699                 79                 42                193 
#            schwann             t_cell 
#                 25                  7</code></pre>
<p>Some of the cell types occur in only a very small number of
cells.</p>
<p>The “size factors” (here, total counts per cell) vary across a very
wide range:</p>
<pre class="r"><code>s &lt;- rowSums(counts)
pdat &lt;- data.frame(log_size_factor = log10(s))
ggplot(pdat,aes(log_size_factor)) +
  geom_histogram(bins = 64,col = &quot;black&quot;,fill = &quot;black&quot;) +
  labs(x = &quot;log(size factor)&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas.Rmd/size-factors-1.png" width="420" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-size-factors-1">
Past versions of size-factors-1.png
</button>
</p>
<div id="fig-size-factors-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/833a14700f5838094e0086f2d237026a880103c7/docs/figure/pancreas.Rmd/size-factors-1.png" target="_blank">833a147</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/fed31e6dbf128b29037d0085f9e21eb008ea58c8/docs/figure/pancreas.Rmd/size-factors-1.png" target="_blank">fed31e6</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Most genes are expressed in at least one cell:</p>
<pre class="r"><code>p &lt;- colSums(counts)/sum(s)
sum(p &gt; 0)
# [1] 18771</code></pre>
<p>The (relative) gene expression levels also vary across a very wide
range:</p>
<pre class="r"><code>p &lt;- p[p &gt; 0]
pdat &lt;- data.frame(log_rel_expression_level = log10(p))
ggplot(pdat,aes(log_rel_expression_level)) +
  geom_histogram(bins = 64,col = &quot;black&quot;,fill = &quot;black&quot;) +
  labs(x = &quot;log-expression level (relative)&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas.Rmd/gene-expression-levels-2-1.png" width="420" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-gene-expression-levels-2-1">
Past versions of gene-expression-levels-2-1.png
</button>
</p>
<div id="fig-gene-expression-levels-2-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/833a14700f5838094e0086f2d237026a880103c7/docs/figure/pancreas.Rmd/gene-expression-levels-2-1.png" target="_blank">833a147</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/fed31e6dbf128b29037d0085f9e21eb008ea58c8/docs/figure/pancreas.Rmd/gene-expression-levels-2-1.png" target="_blank">fed31e6</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Let’s now generate a 2-d nonlinear embedding of the cells using
<em>t</em>-SNE. First, transform the counts into “shifted log
counts”:</p>
<pre class="r"><code>a &lt;- 1
s &lt;- rowSums(counts)
s &lt;- s/mean(s)
Y &lt;- MatrixExtra::mapSparse(counts/(a*s),log1p)</code></pre>
<p>Next, project the cells onto the top 50 PCs:</p>
<pre class="r"><code>set.seed(1)
U &lt;- rsvd(Y,k = 50)$u</code></pre>
<p>Now run t-SNE on the 50 PCs:</p>
<pre class="r"><code>tsne &lt;- Rtsne(U,dims = 2,perplexity = 100,pca = FALSE,
              num_threads = 8,verbose = TRUE)
sample_info$tsne1 &lt;- tsne$Y[,1]
sample_info$tsne2 &lt;- tsne$Y[,2]</code></pre>
<p><em>t</em>-SNE with cells colored by cell-type:</p>
<pre class="r"><code>tsne_colors &lt;- rep(c(&quot;#E69F00&quot;,&quot;#56B4E9&quot;,&quot;#009E73&quot;,&quot;#F0E442&quot;,
                     &quot;#0072B2&quot;,&quot;#D55E00&quot;,&quot;#CC79A7&quot;),times = 2)
tsne_shapes &lt;- rep(c(19,17),each = 7)
ggplot(sample_info,aes(x = tsne1,y = tsne2,color = celltype,
                       shape = celltype)) +
  geom_point(size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  labs(x = &quot;tSNE 1&quot;,y = &quot;tSNE 2&quot;) + 
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas.Rmd/umap-vs-cell-type-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-vs-cell-type-1">
Past versions of umap-vs-cell-type-1.png
</button>
</p>
<div id="fig-umap-vs-cell-type-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/108b85391487fde70d4c859ce2e33b4bebe6e6ce/docs/figure/pancreas.Rmd/umap-vs-cell-type-1.png" target="_blank">108b853</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/833a14700f5838094e0086f2d237026a880103c7/docs/figure/pancreas.Rmd/umap-vs-cell-type-1.png" target="_blank">833a147</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/25f389a7a147366e9fb9f71c43de7664c5e03611/docs/figure/pancreas.Rmd/umap-vs-cell-type-1.png" target="_blank">25f389a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p><em>t</em>-SNE with cells colored by batch:</p>
<pre class="r"><code>ggplot(sample_info,aes(x = tsne1,y = tsne2,color = tech,shape = tech)) +
  geom_point(size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  labs(x = &quot;tSNE 1&quot;,y = &quot;tSNE 2&quot;) + 
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas.Rmd/umap-vs-batch-1.png" width="570" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-vs-batch-1">
Past versions of umap-vs-batch-1.png
</button>
</p>
<div id="fig-umap-vs-batch-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/108b85391487fde70d4c859ce2e33b4bebe6e6ce/docs/figure/pancreas.Rmd/umap-vs-batch-1.png" target="_blank">108b853</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/833a14700f5838094e0086f2d237026a880103c7/docs/figure/pancreas.Rmd/umap-vs-batch-1.png" target="_blank">833a147</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/25f389a7a147366e9fb9f71c43de7664c5e03611/docs/figure/pancreas.Rmd/umap-vs-batch-1.png" target="_blank">25f389a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-10-31
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It is clear from these <em>t</em>-SNE plots that both batch and
cell-type contribute to structure in the data.</p>
<p>For comparison, let’s run the default <em>t</em>-SNE pipeline in
Seurat:</p>
<pre class="r"><code>pancreas &lt;- CreateSeuratObject(counts = t(counts),project = &quot;pancreas&quot;,
                               meta.data = sample_info)
pancreas &lt;- NormalizeData(pancreas)
pancreas &lt;- ScaleData(pancreas)
pancreas &lt;- FindVariableFeatures(pancreas)
pancreas &lt;- RunPCA(pancreas,npcs = 50,features = VariableFeatures(pancreas))
pancreas &lt;- RunTSNE(pancreas)</code></pre>
<p>Seurat <em>t</em>-SNE with cells colored by cell-type:</p>
<pre class="r"><code>DimPlot(pancreas,reduction = &quot;tsne&quot;,group.by = &quot;celltype&quot;,
        shape.by = &quot;celltype&quot;,pt.size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  theme_cowplot(font_size = 10) +
  labs(title = &quot;&quot;)</code></pre>
<p><img src="figure/pancreas.Rmd/seurat-2-1.png" width="600" style="display: block; margin: auto;" /></p>
<p>Seurat <em>t</em>-SNE with cells colored by batch:</p>
<pre class="r"><code>DimPlot(pancreas,reduction = &quot;tsne&quot;,group.by = &quot;tech&quot;,
        shape.by = &quot;tech&quot;,pt.size = 1.5) +
  scale_color_manual(values = tsne_colors) +
  scale_shape_manual(values = tsne_shapes) +
  theme_cowplot(font_size = 10) +
  labs(title = &quot;&quot;)</code></pre>
<p><img src="figure/pancreas.Rmd/seurat-3-1.png" width="540" style="display: block; margin: auto;" /></p>
<p>The default Seurat <em>t</em>-SNE shows does not show as much
structure in the data. This plot does not seem to pick up much batch
structure data; on the other hand, it also does not pick up some of the
more subtyle cell types.</p>
<p>(Note that I had to spend some time customizing the plots; among
other things, Seurat appears to be using the default ggplot color scheme
which is terrible.)</p>
<p>Finally, save the data and t-SNE results to an .Rdata file for more
convenient analysis in R:</p>
<pre class="r"><code>save(list = c(&quot;sample_info&quot;,&quot;counts&quot;),file = &quot;pancreas.RData&quot;)
resaveRdaFiles(&quot;pancreas.RData&quot;)</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS Sonoma 14.6.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] tools     stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] Seurat_5.0.3       SeuratObject_5.0.1 sp_2.1-4           cowplot_1.1.3     
# [5] ggplot2_3.5.0      Rtsne_0.17         rsvd_1.0.5         hdf5r_1.3.11      
# [9] Matrix_1.6-5      
# 
# loaded via a namespace (and not attached):
#   [1] RColorBrewer_1.1-3     jsonlite_1.8.8         magrittr_2.0.3        
#   [4] spatstat.utils_3.0-4   farver_2.1.1           rmarkdown_2.26        
#   [7] fs_1.6.3               vctrs_0.6.5            ROCR_1.0-11           
#  [10] spatstat.explore_3.2-7 htmltools_0.5.7        sass_0.4.8            
#  [13] sctransform_0.4.1      parallelly_1.37.1      KernSmooth_2.23-22    
#  [16] bslib_0.6.1            htmlwidgets_1.6.4      ica_1.0-3             
#  [19] plyr_1.8.9             plotly_4.10.4          zoo_1.8-12            
#  [22] cachem_1.0.8           whisker_0.4.1          igraph_2.0.3          
#  [25] mime_0.12              lifecycle_1.0.4        pkgconfig_2.0.3       
#  [28] R6_2.5.1               fastmap_1.1.1          fitdistrplus_1.1-11   
#  [31] future_1.33.2          shiny_1.8.0            digest_0.6.34         
#  [34] colorspace_2.1-0       patchwork_1.2.0        rprojroot_2.0.4       
#  [37] tensor_1.5             RSpectra_0.16-1        irlba_2.3.5.1         
#  [40] labeling_0.4.3         progressr_0.14.0       fansi_1.0.6           
#  [43] spatstat.sparse_3.0-3  httr_1.4.7             polyclip_1.10-6       
#  [46] abind_1.4-5            compiler_4.3.3         bit64_4.0.5           
#  [49] withr_3.0.0            fastDummies_1.7.3      highr_0.10            
#  [52] float_0.3-2            MASS_7.3-60.0.1        lmtest_0.9-40         
#  [55] httpuv_1.6.14          future.apply_1.11.2    goftest_1.2-3         
#  [58] glue_1.7.0             nlme_3.1-164           promises_1.2.1        
#  [61] grid_4.3.3             cluster_2.1.6          reshape2_1.4.4        
#  [64] generics_0.1.3         gtable_0.3.4           spatstat.data_3.0-4   
#  [67] tidyr_1.3.1            data.table_1.15.2      utf8_1.2.4            
#  [70] spatstat.geom_3.2-9    RcppAnnoy_0.0.22       ggrepel_0.9.5         
#  [73] RANN_2.6.1             pillar_1.9.0           stringr_1.5.1         
#  [76] spam_2.10-0            RcppHNSW_0.6.0         later_1.3.2           
#  [79] splines_4.3.3          dplyr_1.1.4            lattice_0.22-5        
#  [82] survival_3.5-8         bit_4.0.5              deldir_2.0-4          
#  [85] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         
#  [88] knitr_1.45             git2r_0.33.0           gridExtra_2.3         
#  [91] scattermore_1.2        RhpcBLASctl_0.23-42    xfun_0.42             
#  [94] matrixStats_1.2.0      stringi_1.8.3          workflowr_1.7.1       
#  [97] lazyeval_0.2.2         yaml_2.3.8             evaluate_0.23         
# [100] codetools_0.2-19       tibble_3.2.1           cli_3.6.2             
# [103] uwot_0.2.2.9000        xtable_1.8-4           reticulate_1.36.1     
# [106] munsell_0.5.0          jquerylib_0.1.4        Rcpp_1.0.12           
# [109] globals_0.16.3         spatstat.random_3.2-3  png_0.1-8             
# [112] parallel_4.3.3         ellipsis_0.3.2         dotCall64_1.1-1       
# [115] listenv_0.9.1          viridisLite_0.4.2      MatrixExtra_0.1.15    
# [118] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        
# [121] purrr_1.0.2            rlang_1.1.3</code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
