<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>Various attempts at factorizing the pancreas data set</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Various attempts at factorizing the
pancreas data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-01-29
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreebb2aac59f2fd2ef0b917cf474acd0f7125ecae66targetblankbb2aac5a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/bb2aac59f2fd2ef0b917cf474acd0f7125ecae66" target="_blank">bb2aac5</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetreebb2aac59f2fd2ef0b917cf474acd0f7125ecae66targetblankbb2aac5a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/bb2aac59f2fd2ef0b917cf474acd0f7125ecae66" target="_blank">bb2aac5</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad

Unstaged changes:
    Modified:   code/annotation_plots.R

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pancreas_factors.Rmd</code>) and
HTML (<code>docs/pancreas_factors.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/bb2aac59f2fd2ef0b917cf474acd0f7125ecae66/analysis/pancreas_factors.Rmd" target="_blank">bb2aac5</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE,
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/pancreas_factors.html" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/pancreas_factors.html" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
Improved the structure plots for the semi-NMF results in the
pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/94e20ef56fbcafeef20133fdcc0fa9e9397804d0/analysis/pancreas_factors.Rmd" target="_blank">94e20ef</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/808032e104fda715b179470b0935227be3803eaf/docs/pancreas_factors.html" target="_blank">808032e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
Fixed a couple mistakes in the pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/ace097c631f0cf66f9bbc944d181a33e330c3450/analysis/pancreas_factors.Rmd" target="_blank">ace097c</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/826307e4323c932270ff6224d9e7b4f2223ae36f/docs/pancreas_factors.html" target="_blank">826307e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
Build site.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/0a324298d57885f6d4417faed4f8902a553b41cc/analysis/pancreas_factors.Rmd" target="_blank">0a32429</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/d20a98559d64785db97f32f24b19de4eb45bef97/docs/pancreas_factors.html" target="_blank">d20a985</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
Added some summary text to the beginning of the pancreas_factors
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/b58c81a1cc93d93c927e14cbf02d5179aed24ff2/analysis/pancreas_factors.Rmd" target="_blank">b58c81a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE,
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/pancreas_factors.html" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
A few more improvements to the structure plots in the pancreas_factors
analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/86a4d0983f43f32f0806812b5a9ddd00ab6fb5a2/analysis/pancreas_factors.Rmd" target="_blank">86a4d09</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
workflowr::wflow_publish("pancreas_factors.Rmd", verbose = TRUE,
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a4ae6935472afd420bb6b76408712c8ef7915236/analysis/pancreas_factors.Rmd" target="_blank">a4ae693</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
Added semi-NMF results to pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/a24828a95ade05a61c8c51bf329203bd61918541/analysis/pancreas_factors.Rmd" target="_blank">a24828a</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
Added fastTopics structure plots to pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e9fc660b67f1a30c0111089f2ba303a5f4ebe578/analysis/pancreas_factors.Rmd" target="_blank">e9fc660</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
Added structure plots for nnlm results to pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/b5e911083d472a67a1c7ec3a8d5badbdfd9af5d8/analysis/pancreas_factors.Rmd" target="_blank">b5e9110</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
<td>
Added structure plots for flashier NMF to pancreas_factors analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/3da799758f5e1eddbaff1b85dfa4664502b6ae80/analysis/pancreas_factors.Rmd" target="_blank">3da7997</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-21
</td>
<td>
Added code chunk to pancreas_factors analysis to compare timings.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/b9d08ddcd5b91ceed359f8e9946212d0bb54639a/docs/pancreas_factors.html" target="_blank">b9d08dd</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-21
</td>
<td>
First build of the pancreas_factors workflowr analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/5d5c725a24a3931431ae4a734d42a39875eed893/analysis/pancreas_factors.Rmd" target="_blank">5d5c725</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-21
</td>
<td>
Added pancreas_factors.Rmd.
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>This goal of this analysis is to apply a few different matrix
factorization approaches to the <a href="pancreas.html">pancreas data
set</a> and assess how well they work for this data set.</p>
<p>First, load the packages needed for this analysis.</p>
<pre class="r"><code>library(Matrix)
library(fastTopics)
library(ggplot2)
library(cowplot)</code></pre>
<p>Set the seed for reproducibility.</p>
<pre class="r"><code>set.seed(1)</code></pre>
<p>This is a function I will use below to subsample the cell types that
have a lot of cells so that the rare cell types are more visible in the
Structure plots.</p>
<pre class="r"><code>subsample_cell_types &lt;- function (x, n = 1000) {
  cells &lt;- NULL
  groups &lt;- levels(x)
  for (g in groups) {
    i  &lt;-  which(x == g)
    n0 &lt;- min(n,length(i))
    i  &lt;- sample(i,n0)
    cells &lt;- c(cells,i)
  }
  return(sort(cells))
}</code></pre>
<p>Load the pancreas data set and the outputs generated from running the
<code>compute_pancreas_factors.R</code> and
<code>compute_pancreas_factors2.R</code> scripts.</p>
<pre class="r"><code>load(&quot;../data/pancreas.RData&quot;)
load(&quot;../output/pancreas_factors.RData&quot;)
timings0 &lt;- timings
load(&quot;../output/pancreas_factors2.RData&quot;)
timings &lt;- c(timings0,timings)</code></pre>
<p>This is a summary of the running times (in hours) for the model
fitting (focus on the “elapsed” column):</p>
<pre class="r"><code>round(do.call(rbind,timings)/60^2,digits = 2)
#            user.self sys.self elapsed user.child sys.child
# nmf             0.17     0.02    0.12          0         0
# fl_nmf          8.33     0.02    1.43          0         0
# fl_snmf         8.31     0.02    1.38          0         0
# fasttopics      4.97     0.42    1.81          0         0
# fl_nmf_cc       8.44     0.07    1.50          0         0</code></pre>
<p>Fitting the topic model (“fasttopics”) and the three flashier models
(“NMF”, “semi-NMF”, “NMF-CC”) each took 1-2 hours. Fitting the NMF model
using the NNLM package was much, much faster (about 7 minutes).
<em>Note:</em> Below I’ll explain what is the “NMF-CC” model.</p>
<div id="flashier-nmf" class="section level3">
<h3>flashier NMF</h3>
<p>The empirical Bayes NMF compellingly identifies several factors
corresponding very well to both cell type and sequencing technology:</p>
<pre class="r"><code>cells &lt;- subsample_cell_types(sample_info$celltype,n = 500)
L &lt;- fl_nmf_ldf$L
k &lt;- ncol(L)
colnames(L) &lt;- paste0(&quot;k&quot;,1:k)
batch_factors    &lt;- c(2:5,7:8,20)
celltype_factors &lt;- c(6,11:19,21)
other_factors    &lt;- c(1,9:10,22:23)
celltype &lt;- sample_info$celltype
celltype &lt;-
 factor(celltype,
        c(&quot;acinar&quot;,&quot;ductal&quot;,&quot;activated_stellate&quot;,&quot;quiescent_stellate&quot;,
          &quot;endothelial&quot;,&quot;macrophage&quot;,&quot;mast&quot;,&quot;schwann&quot;,&quot;t_cell&quot;,&quot;alpha&quot;,
          &quot;beta&quot;,&quot;delta&quot;,&quot;gamma&quot;,&quot;epsilon&quot;))
p1 &lt;- structure_plot(L,topics = batch_factors,grouping = sample_info$tech,
                     gap = 3,perplexity = 70,n = 500) +
  labs(y = &quot;membership&quot;,title = &quot;data-set factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p2 &lt;- structure_plot(L[cells,],topics = celltype_factors,
                     grouping = celltype[cells],gap = 3,
                     n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;cell-type factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p3 &lt;- structure_plot(L[cells,],topics = other_factors,
                     grouping = celltype[cells],gap = 3,
                     n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;other factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
plot_grid(p1,p2,p3,nrow = 3,ncol = 1)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-flashier-nmf-1">
Past versions of structure-plots-flashier-nmf-1.png
</button>
</p>
<div id="fig-structure-plots-flashier-nmf-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that the number of factors, 23, was automatically determined by
flashier. Some structure was missed by this factorization and it is
possible that forcing flashier to estimate more factors would have also
picked up other intersting structure. Nonetheless, the flashier NMF
seems to cope well here with separating the cell types from the batch
effects.</p>
</div>
<div id="nmf-nnlm" class="section level3">
<h3>NMF (NNLM)</h3>
<p>To contrast this flashier result with a standard NMF (generated by
NNLM), the factors in the standard NMF does not capture nearly as well
the cell types and the batch structure. We don’t have the ground truth,
but better correspondence between the provided cell annotations and the
flashier result point toward the flashier fit being better.</p>
<pre class="r"><code>scale_cols &lt;- function (A, b)
  t(t(A) * b)
W &lt;- nmf$W
k &lt;- ncol(W)
d &lt;- apply(W,2,max)
W &lt;- scale_cols(W,1/d)
colnames(W) &lt;- paste0(&quot;k&quot;,1:k)
batch_factors    &lt;- c(1,6,7,10,13,14,16:18)
celltype_factors &lt;- c(2,5,8,9,11,12,15,20:23)
other_factors    &lt;- c(3,4,19)
p1 &lt;- structure_plot(W[,batch_factors],grouping = sample_info$tech,
                     gap = 3,perplexity = 70,n = 500) +
  labs(y = &quot;membership&quot;,title = &quot;data-set factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p2 &lt;- structure_plot(W[cells,celltype_factors],grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;cell-type factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p3 &lt;- structure_plot(W[cells,other_factors],grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;other factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
plot_grid(p1,p2,p3,nrow = 3,ncol = 1)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-nnlm-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-nnlm-1">
Past versions of structure-plots-nnlm-1.png
</button>
</p>
<div id="fig-structure-plots-nnlm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-nnlm-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-nnlm-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-nnlm-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>It is interesting that the flashier NMF factors are much more
sparse:</p>
<pre class="r"><code>pdat &lt;- data.frame(flashier = quantile(L,seq(0,1,0.01)),
                   nnlm     = quantile(W,seq(0,1,0.01)))
ggplot(pdat,aes(x = nnlm,y = flashier)) +
  geom_line() +
  geom_point() +
  geom_abline(slope = 1,intercept = 0,color = &quot;magenta&quot;,linetype = &quot;dotted&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/flashier-vs-nnlm-1.png" width="300" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-flashier-vs-nnlm-1">
Past versions of flashier-vs-nnlm-1.png
</button>
</p>
<div id="fig-flashier-vs-nnlm-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/flashier-vs-nnlm-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="topic-model-fasttopics" class="section level3">
<h3>Topic model (fastTopics)</h3>
<p>The topic model (again with 23 factors/topics) has more difficulty
than NMF separating the cell types from the batch/data-set effects:</p>
<pre class="r"><code>L &lt;- poisson2multinom(pnmf)$L
batch_topics     &lt;- c(2,5,6,11,12)
celltype_topics  &lt;- c(4,9,8,15,16,17,18,19,20)
celltype_topics2 &lt;- c(1,3,7,10,13,14,21,22,23)
p1 &lt;- structure_plot(L[,batch_topics],grouping = sample_info[,&quot;tech&quot;],
                     gap = 3,perplexity = 70,n = 500) +
  ggtitle(&quot;data-set topics&quot;)
p2 &lt;- structure_plot(L[cells,celltype_topics],grouping = celltype[cells],
                     gap = 3,perplexity = 70,n = 500) +
  ggtitle(&quot;cell-type topics&quot;)                    
p3 &lt;- structure_plot(L[cells,celltype_topics2],grouping = celltype[cells],
                     gap = 3,perplexity = 70,n = 500) +
  ggtitle(&quot;more cell-type topics&quot;)                   
plot_grid(p1,p2,p3,nrow = 3,ncol = 1)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-fasttopics-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-fasttopics-1">
Past versions of structure-plots-fasttopics-1.png
</button>
</p>
<div id="fig-structure-plots-fasttopics-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>For example, topics 3, 10 and 23 have identified cell types specific
to the Fluidigm C1 data only, whereas it would be preferrable to
separate these data-set effects from the cell types:</p>
<pre class="r"><code>i &lt;- which(sample_info$tech == &quot;fluidigmc1&quot;)
structure_plot(L[i,celltype_topics2],grouping = celltype[i],
               gap = 3,perplexity = 30,n = 500)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-fasttopics-fluidicmg1-1.png" width="720" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-fasttopics-fluidicmg1-1">
Past versions of structure-plots-fasttopics-fluidicmg1-1.png
</button>
</p>
<div id="fig-structure-plots-fasttopics-fluidicmg1-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-fluidicmg1-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-fluidicmg1-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-fasttopics-fluidicmg1-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
</div>
<div id="flashier-semi-nmf" class="section level3">
<h3>Flashier semi-NMF</h3>
<p>The semi-NMF fit generated by flashier seems promising in that it
separates data-set structure from cell-type structure. The individual
factors seem at first glance to be harder to interpret than the flashier
NMF factors, but the results nonetheless are intruiging:</p>
<pre class="r"><code>L &lt;- fl_snmf_ldf$L
x &lt;- apply(L,2,function (x) quantile(x,0.995))
L &lt;- scale_cols(L,1/x)
colnames(L) &lt;- paste0(&quot;k&quot;,1:k)
batch_factors    &lt;- c(4:6,11,22:23)
celltype_factors &lt;- c(3,7:10,12,17:18,20,21)
other_factors    &lt;- c(1:2,13:16,19)
p1 &lt;- structure_plot(L[,batch_factors],grouping = sample_info$tech,gap = 3,
                     perplexity = 70,n = 500) +
  labs(y = &quot;membership&quot;,title = &quot;data-set factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p2 &lt;- structure_plot(L[cells,celltype_factors],
                     grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;cell-type factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p3 &lt;- structure_plot(L[cells,other_factors],
                     grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;other factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
plot_grid(p1 + ylim(0,3),
          p2 + ylim(0,4),
          p3 + ylim(0,4),
          nrow = 3,ncol = 1,
          rel_heights = c(2,3,3))</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-flashier-snmf-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-flashier-snmf-1">
Past versions of structure-plots-flashier-snmf-1.png
</button>
</p>
<div id="fig-structure-plots-flashier-snmf-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-snmf-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-snmf-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-snmf-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>I found it more difficult to create nice Structure plots from the
semi-NMF result because of the “spikiness” of the loadings.</p>
</div>
<div id="flashier-nmf-with-cross-cutting-factors"
class="section level3">
<h3>flashier NMF with “cross-cutting factors”</h3>
<p>Finally, mainly to see whether it was possible, I also tried fitting
an NMF using flashier in which there were additional “cross-cutting
factors” (CC) that could capture both increases and decreases in
expression, with the hope that these CC factors could better capture the
data-set effects. It is unclear this provided an improvement over the
flashier NMF results (which were already quite good), but I think it is
intruiging this is a possibility in flashier.</p>
<pre class="r"><code>set.seed(1)
L &lt;- fl_nmf_cc_ldf$L
k &lt;- ncol(L)
colnames(L) &lt;- paste0(&quot;k&quot;,1:k)
batch_factors    &lt;- c(1,2:8)
celltype_factors &lt;- c(9,12,13:21)
other_factors    &lt;- c(10,11,22:23)
p1 &lt;- structure_plot(L,topics = batch_factors,grouping = sample_info$tech,
                     gap = 3,perplexity = 70, n = 500) +
  labs(y = &quot;membership&quot;,title = &quot;data-set factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p2 &lt;- structure_plot(L[cells,],topics = celltype_factors,
                     grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;cell-type factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
p3 &lt;- structure_plot(L[cells,],topics = other_factors,
                     grouping = celltype[cells],
                     gap = 3,n = 500,perplexity = 70) +
  labs(y = &quot;membership&quot;,title = &quot;other factors&quot;,
       fill = &quot;factor&quot;,color = &quot;factor&quot;)
plot_grid(p1,p2,p3,nrow = 3,ncol = 1)</code></pre>
<p><img src="figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-cc-1.png" width="960" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-structure-plots-flashier-nmf-cc-1">
Past versions of structure-plots-flashier-nmf-cc-1.png
</button>
</p>
<div id="fig-structure-plots-flashier-nmf-cc-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8ccf088c50dc44f2f55ba3e846347f9f1eb6bd03/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-cc-1.png" target="_blank">8ccf088</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-29
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/e02c1f77d75df1f588d88fd90306da9bac9d2016/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-cc-1.png" target="_blank">e02c1f7</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-27
</td>
</tr>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/10bdb89dc9d8b624babed358b09ad88f76aa9a17/docs/figure/pancreas_factors.Rmd/structure-plots-flashier-nmf-cc-1.png" target="_blank">10bdb89</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2024-11-26
</td>
</tr>
</tbody>
</table>
</div>
</div>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS Sonoma 14.7.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] stats     graphics  grDevices utils     datasets  methods   base     
# 
# other attached packages:
# [1] cowplot_1.1.3     ggplot2_3.5.0     fastTopics_0.7-07 Matrix_1.6-5     
# 
# loaded via a namespace (and not attached):
#  [1] gtable_0.3.4        xfun_0.42           bslib_0.6.1        
#  [4] htmlwidgets_1.6.4   ggrepel_0.9.5       lattice_0.22-5     
#  [7] quadprog_1.5-8      vctrs_0.6.5         tools_4.3.3        
# [10] generics_0.1.3      parallel_4.3.3      tibble_3.2.1       
# [13] fansi_1.0.6         highr_0.10          pkgconfig_2.0.3    
# [16] data.table_1.15.2   SQUAREM_2021.1      RcppParallel_5.1.7 
# [19] lifecycle_1.0.4     truncnorm_1.0-9     farver_2.1.1       
# [22] compiler_4.3.3      stringr_1.5.1       git2r_0.33.0       
# [25] progress_1.2.3      munsell_0.5.0       RhpcBLASctl_0.23-42
# [28] httpuv_1.6.14       htmltools_0.5.7     sass_0.4.8         
# [31] yaml_2.3.8          lazyeval_0.2.2      plotly_4.10.4      
# [34] crayon_1.5.2        later_1.3.2         pillar_1.9.0       
# [37] jquerylib_0.1.4     whisker_0.4.1       tidyr_1.3.1        
# [40] uwot_0.2.2.9000     cachem_1.0.8        gtools_3.9.5       
# [43] tidyselect_1.2.1    digest_0.6.34       Rtsne_0.17         
# [46] stringi_1.8.3       dplyr_1.1.4         purrr_1.0.2        
# [49] ashr_2.2-66         labeling_0.4.3      rprojroot_2.0.4    
# [52] fastmap_1.1.1       grid_4.3.3          colorspace_2.1-0   
# [55] cli_3.6.2           invgamma_1.1        magrittr_2.0.3     
# [58] utf8_1.2.4          withr_3.0.0         prettyunits_1.2.0  
# [61] scales_1.3.0        promises_1.2.1      rmarkdown_2.26     
# [64] httr_1.4.7          workflowr_1.7.1     hms_1.1.3          
# [67] pbapply_1.7-2       evaluate_0.23       knitr_1.45         
# [70] viridisLite_0.4.2   irlba_2.3.5.1       rlang_1.1.3        
# [73] Rcpp_1.0.12         mixsqp_0.3-54       glue_1.7.0         
# [76] jsonlite_1.8.8      R6_2.5.1            fs_1.6.3</code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
