<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Peter Carbonetto" />


<title>Initial exploration of the pancreas endocrine data set</title>

<script src="site_libs/header-attrs-2.26/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/readable.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark the anchor link active (and if it's in a dropdown, also mark that active)
  var dropdown = menuAnchor.closest('li.dropdown');
  if (window.bootstrap) { // Bootstrap 4+
    menuAnchor.addClass('active');
    dropdown.find('> .dropdown-toggle').addClass('active');
  } else { // Bootstrap 3
    menuAnchor.parent().addClass('active');
    dropdown.addClass('active');
  }

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before, .tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "\e259";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "\e258";
  font-family: 'Glyphicons Halflings';
  border: none;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->




</head>

<body>


<div class="container-fluid main-container">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">single-cell jamboree</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Overview</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/stephenslab/single-cell-jamboree">source</a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">Initial exploration of the pancreas
endocrine data set</h1>
<h4 class="author">Peter Carbonetto</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2025-01-09
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong>
<code>single-cell-jamboree/analysis/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.1). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed1code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(1)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed1code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(1)</code> was run prior to running the
code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree38cba6525f74db21aa5f73664165426c0e2ffa6atargetblank38cba65a">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong>
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/38cba6525f74db21aa5f73664165426c0e2ffa6a" target="_blank">38cba65</a>
</a>
</p>
</div>
<div
id="strongRepositoryversionstrongahrefhttpsgithubcomstephenslabsinglecelljamboreetree38cba6525f74db21aa5f73664165426c0e2ffa6atargetblank38cba65a"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version
<a href="https://github.com/stephenslab/single-cell-jamboree/tree/38cba6525f74db21aa5f73664165426c0e2ffa6a" target="_blank">38cba65</a>.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
Untracked files:
    Untracked:  data/GSE132188_adata.h5ad.h5
    Untracked:  data/Immune_ALL_human.h5ad
    Untracked:  data/pancreas_endocrine.RData
    Untracked:  data/pancreas_endocrine_alldays.h5ad

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/pancreas_endocrine.Rmd</code>)
and HTML (<code>docs/pancreas_endocrine.html</code>) files. If you’ve
configured a remote Git repository (see <code>?wflow_git_remote</code>),
click on the hyperlinks in the table below to view the files as they
were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/38cba6525f74db21aa5f73664165426c0e2ffa6a/analysis/pancreas_endocrine.Rmd" target="_blank">38cba65</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
wflow_publish("pancreas_endocrine.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/8a7e118370813930efd29f3b1e759b4e0198f13a/docs/pancreas_endocrine.html" target="_blank">8a7e118</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
Added some umap plots to the pancreas_endocrine analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/f5bb283ecf4c2ebebb1f493a362ed21adf0bae1d/analysis/pancreas_endocrine.Rmd" target="_blank">f5bb283</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
wflow_publish("pancreas_endocrine.Rmd", verbose = TRUE, view = FALSE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/9badd1679987e0fdf4c89a0351c800bdfafaceff/analysis/pancreas_endocrine.Rmd" target="_blank">9badd16</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
Made a few improvements to the pancreas_endocrine analysis.
</td>
</tr>
<tr>
<td>
html
</td>
<td>
<a href="https://rawcdn.githack.com/stephenslab/single-cell-jamboree/19b7131c95bc40335d9a1a9eadbb77c1cb85e53a/docs/pancreas_endocrine.html" target="_blank">19b7131</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
First build of the pancreas_endocrine analysis.
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/0aad00e53f277b4ab71fc02e9ab37007e5c89087/analysis/pancreas_endocrine.Rmd" target="_blank">0aad00e</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
workflowr::wflow_publish("pancreas_endocrine.Rmd", verbose = TRUE)
</td>
</tr>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/f900873ec820b5dabb81f76b6a68a18428c51597/analysis/pancreas_endocrine.Rmd" target="_blank">f900873</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
<td>
workflowr::wflow_publish("index.Rmd")
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<p>The aim of this analysis is to take an initial look at the mouse
pancreas endocrinogenesis data from <a
href="https://doi.org/10.1242/dev.173849">Bastidas-Ponce et al 2019</a>
(see also <a
href="https://github.com/theislab/pancreatic-endocrinogenesis">this
GitHub repository</a>) and that was later analyzed in the <a
href="https://doi.org/10.1038/s41587-020-0591-3">scVelo paper</a>.</p>
<p>To run the code, you will need to first download the
“GSE132188_adata.h5ad.h5” file from the GEO website, accession
GSE132188.</p>
<p>Then run the Python script
<code>prepare_pancreas_endocrine_data.py</code> to generate the file
“pancreas_endocrine_alldays.h5ad”.</p>
<p>First, load the packages needed for this analysis.</p>
<pre class="r"><code>library(Matrix)
library(anndata)
library(reticulate)
library(tools)
library(ggplot2)
library(cowplot)
Sys.getenv(&quot;RETICULATE_PYTHON&quot;)
# [1] &quot;/Users/pcarbo/miniforge3/bin/python&quot;</code></pre>
<p><strong>Note:</strong> The AnnData Python package is needed to run
this code. I installed anndata 0.11.2 using conda.</p>
<p>Get the count data that was prepared using the Python script.</p>
<pre class="r"><code>dat1   &lt;- read_h5ad(&quot;../data/pancreas_endocrine_alldays.h5ad&quot;)
counts &lt;- dat1$X
counts &lt;- as(counts,&quot;CsparseMatrix&quot;)</code></pre>
<p>Get the meta-data downloaded from GEO.</p>
<pre class="r"><code>dat2 &lt;- read_h5ad(&quot;../data/GSE132188_adata.h5ad.h5&quot;)</code></pre>
<p>Align the two data sets.</p>
<pre class="r"><code>ids1   &lt;- rownames(dat1$obs)
ids2   &lt;- rownames(dat2$obs)
ids2   &lt;- paste0(&quot;e&quot;,10*as.numeric(as.character(dat2$obs$day)),&quot;-&quot;,ids2)
ids2   &lt;- substr(ids2,1,23)
rows   &lt;- which(is.element(ids1,ids2))
ids1   &lt;- ids1[rows]
counts &lt;- counts[rows,]
obs1   &lt;- dat1$obs[rows,]</code></pre>
<p>Check that the sample ids and genes are the same.</p>
<pre class="r"><code>print(all(ids1 == ids2))
print(all(rownames(dat1$var) == rownames(dat2$var)))
# [1] TRUE
# [1] TRUE</code></pre>
<p>Extract the gene info.</p>
<pre class="r"><code>gene_info &lt;- dat2$var
gene_info &lt;- cbind(gene = rownames(gene_info),gene_info)
rownames(gene_info) &lt;- NULL</code></pre>
<p>Extract the sample info.</p>
<pre class="r"><code>sample_info &lt;- dat2$obs
umap &lt;- dat2$obsm$X_umap
colnames(umap) &lt;- c(&quot;umap1&quot;,&quot;umap2&quot;)
sample_info &lt;- cbind(data.frame(id = ids1,stringsAsFactors = FALSE),
                     umap,
                     sample_info)
rownames(sample_info) &lt;- NULL</code></pre>
<p>Save the data and t-SNE results to an .Rdata file for more convenient
analysis in R:</p>
<pre class="r"><code>save(list = c(&quot;gene_info&quot;,&quot;sample_info&quot;,&quot;counts&quot;),
     file = &quot;pancreas_endocrine.RData&quot;)
resaveRdaFiles(&quot;pancreas_endocrine.RData&quot;)</code></pre>
<p>The meta-data includes a previously computed UMAP which we can use to
visualize of the key structure in the data.</p>
<p>The Bastidas-Ponce et al paper identified 8 main cell types:</p>
<pre class="r"><code>cluster_colors &lt;- c(&quot;#e41a1c&quot;,&quot;#377eb8&quot;,&quot;#4daf4a&quot;,&quot;#984ea3&quot;,&quot;#ff7f00&quot;,
                    &quot;#ffff33&quot;,&quot;#a65628&quot;,&quot;#f781bf&quot;)
ggplot(sample_info,
       aes(x = umap1,y = umap2,color = clusters_fig2_final)) +
  geom_point(size = 0.75) +
  scale_color_manual(values = cluster_colors) +
  labs(color = &quot;cluster&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_endocrine.Rmd/umap-cell-types-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-cell-types-1">
Past versions of umap-cell-types-1.png
</button>
</p>
<div id="fig-umap-cell-types-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8a7e118370813930efd29f3b1e759b4e0198f13a/docs/figure/pancreas_endocrine.Rmd/umap-cell-types-1.png" target="_blank">8a7e118</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Additionally, they distinguished proliferating vs. non-proliferating
cells:</p>
<pre class="r"><code>ggplot(sample_info,
       aes(x = umap1,y = umap2,color = proliferation)) +
  geom_point(size = 0.75) +
  scale_color_manual(values = c(&quot;dodgerblue&quot;,&quot;darkblue&quot;)) +
  labs(color = &quot;&quot;) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_endocrine.Rmd/umap-proliferating-1.png" width="480" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-proliferating-1">
Past versions of umap-proliferating-1.png
</button>
</p>
<div id="fig-umap-proliferating-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8a7e118370813930efd29f3b1e759b4e0198f13a/docs/figure/pancreas_endocrine.Rmd/umap-proliferating-1.png" target="_blank">8a7e118</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Beyond the main cell types and proliferating/non-profilerating, there
appears to be additional structure in the data corresponding to the
different lineages (days):</p>
<pre class="r"><code>lineage_colors &lt;- c(&quot;#d01c8b&quot;,&quot;#f1b6da&quot;,&quot;#b8e186&quot;,&quot;#4dac26&quot;)
ggplot(sample_info,
       aes(x = umap1,y = umap2,color = day)) +
  geom_point(size = 0.75) +
  scale_color_manual(values = lineage_colors) +
  theme_cowplot(font_size = 10)</code></pre>
<p><img src="figure/pancreas_endocrine.Rmd/umap-lineage-1.png" width="450" style="display: block; margin: auto;" /></p>
<p>
<button type="button" class="btn btn-default btn-xs btn-workflowr btn-workflowr-fig" data-toggle="collapse" data-target="#fig-umap-lineage-1">
Past versions of umap-lineage-1.png
</button>
</p>
<div id="fig-umap-lineage-1" class="collapse">
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
<a href="https://github.com/stephenslab/single-cell-jamboree/blob/8a7e118370813930efd29f3b1e759b4e0198f13a/docs/figure/pancreas_endocrine.Rmd/umap-lineage-1.png" target="_blank">8a7e118</a>
</td>
<td>
Peter Carbonetto
</td>
<td>
2025-01-09
</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>Note that these UMAPs plots reproduce the plots in <a
href="https://doi.org/10.1242/dev.173849">the original paper</a>.</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()
# R version 4.3.3 (2024-02-29)
# Platform: aarch64-apple-darwin20 (64-bit)
# Running under: macOS Sonoma 14.7.1
# 
# Matrix products: default
# BLAS:   /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRblas.0.dylib 
# LAPACK: /Library/Frameworks/R.framework/Versions/4.3-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.11.0
# 
# locale:
# [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
# 
# time zone: America/Chicago
# tzcode source: internal
# 
# attached base packages:
# [1] tools     stats     graphics  grDevices utils     datasets  methods  
# [8] base     
# 
# other attached packages:
# [1] cowplot_1.1.3     ggplot2_3.5.0     reticulate_1.40.0 anndata_0.7.5.6  
# [5] Matrix_1.6-5     
# 
# loaded via a namespace (and not attached):
#  [1] sass_0.4.8       utf8_1.2.4       generics_0.1.3   stringi_1.8.3   
#  [5] lattice_0.22-5   digest_0.6.34    magrittr_2.0.3   evaluate_0.23   
#  [9] grid_4.3.3       fastmap_1.1.1    rprojroot_2.0.4  workflowr_1.7.1 
# [13] jsonlite_1.8.8   whisker_0.4.1    promises_1.2.1   fansi_1.0.6     
# [17] scales_1.3.0     jquerylib_0.1.4  cli_3.6.2        rlang_1.1.3     
# [21] munsell_0.5.0    withr_3.0.0      cachem_1.0.8     yaml_2.3.8      
# [25] dplyr_1.1.4      colorspace_2.1-0 httpuv_1.6.14    assertthat_0.2.1
# [29] vctrs_0.6.5      R6_2.5.1         png_0.1-8        lifecycle_1.0.4 
# [33] git2r_0.33.0     stringr_1.5.1    fs_1.6.3         pkgconfig_2.0.3 
# [37] pillar_1.9.0     bslib_0.6.1      later_1.3.2      gtable_0.3.4    
# [41] glue_1.7.0       Rcpp_1.0.12      highr_0.10       xfun_0.42       
# [45] tibble_3.2.1     tidyselect_1.2.1 knitr_1.45       farver_2.1.1    
# [49] htmltools_0.5.7  rmarkdown_2.26   labeling_0.4.3   compiler_4.3.3</code></pre>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>





</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
